<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Candida Migratory TCR Project</title>
    <link rel="stylesheet" href="styles.css">
</head>

<body>
    <header>
        <h1>Candida Migratory TCR Project</h1>
        <nav>
            <ul>
                <li><a href="/Candida_Migratory_TCR/contact.html">Contact</a></li>
                <li><a href="#project-introduction">Project Introduction</a></li>
                <li><a href="#r-scripts">R Scripts</a></li>
                <li><a href="#tcr-scripts">TCR Scripts</a></li>
                <li><a href="#python-scripts">Python Scripts</a></li>
                <li><a href="#data">Data</a></li>
                <li><a href="#figures-tables">Figures & Tables</a></li>
                <li><a href="#tcr-packages">TCR Packages</a></li>
            </ul>
        </nav>
    </header>

    <section id="project-introduction">
        <h2>Project Introduction</h2>
        <p>.</p>
    </section>

    <section id="r-scripts">
        <h2>R Scripts for scRNA/Cite-Seq Analysis</h2>
        <p>Scripts related to scRNA analysis and clustering.</p>
        <ul>
            <li>
                <a href="/Candida_Migratory_TCR/Scripts/R/sc_RNA_scripts/Candida_sc_QC.Rmd">Candida_sc_QC.Rmd</a>
                - Quality control for single-cell RNA-seq and Cite-seq data
                - Importing, removing outlier cells, demultiplexing
                - Pre-QC Thresholds - <a href="/Candida_Migratory_TCR/Figures_Tables/supplemental_1.jpeg">supplemental_1.jpeg</a>
                - Post-QC Thresholds - <a href="/Candida_Migratory_TCR/Figures_Tables/supplemental_2.jpeg">supplemental_2.jpeg</a>
            </li>
            <li>
                <a href="/Candida_Migratory_TCR/Scripts/R/sc_RNA_scripts/Candida_sc_CD4_CD8.Rmd">Candida_sc_CD4_CD8.Rmd</a>
                - Splitting our cells into CD4 and CD8 groups
                - Dimension reduction, clustering, organ distribution per cluster, visualizing marker genes
                - Exploration of double negative (CD4-/CD8- cells) and their distribution per cluster
                - Figure 1: CD4/CD8 Split - <a href="/Candida_Migratory_TCR/Figures_Tables/figure_1.jpeg">figure_1.jpeg</a>
                - Figure 2: Double Negative Distribution  - <a href="/Candida_Migratory_TCR/Figures_Tables/figure_2.jpeg">figure_2.jpeg</a>
                - Figure 3: Organ Distribution  - <a href="/Candida_Migratory_TCR/Figures_Tables/figure_3.jpeg">figure_3.jpeg</a>
                - Dimension Reduction Elbow Plots and PCA - <a href="/Candida_Migratory_TCR/Figures_Tables/supplemental_3.jpeg">supplemental_3.jpeg</a>
                
                
                - Analysis of single-cell data, including dimension reduction, clustering, differential abundance analysis, identifying marker genes, cell type annotations, pseudotime analysis and cell-signaling analysis  (date: 11/02/2023).
            </li>
            <li>
                <a href="/Candida_Migratory_TCR/Scripts/R/sc_RNA_scripts/Candida_sc_diff_exp_&_annot_CD4CD8.Rmd">Candida_sc_diff_exp_&_annot_CD4CD8.Rmd</a>
                <ul>
                    <li>- Differential expression analysis (finding DEGs/marker gene expression, protein expression), cell type annotations</li>
                    <li>- Gene ontology and pathway analysis</li>
                    <li>- Pseudobulking for Visualization</li>

                    <li>- CD4:</li>
                    <li>- Figure 4: CD4 Cluster Visualization and Expression - <a href="/Candida_Migratory_TCR/Figures_Tables/figure_4.jpeg">figure_4.jpeg</a></li>
                    <li>- CD4 Volcano Plots - <a href="/Candida_Migratory_TCR/Figures_Tables/supplemental_6.jpeg">supplemental_6.jpeg</a></li>
                    <li>- CD4 Cell Cycle Score - <a href="/Candida_Migratory_TCR/Figures_Tables/supplemental_7.jpeg">supplemental_7.jpeg</a></li>
                    <li>- Figure 6: CD4 Gene Ontology Analysis - <a href="/Candida_Migratory_TCR/Figures_Tables/figure_6.jpeg">figure_6.jpeg</a></li>

                    <li>- CD8:</li>
                    <li>- Figure 5: CD8 Cluster Visualization and Expression - <a href="/Candida_Migratory_TCR/Figures_Tables/figure_5.jpeg">figure_5.jpeg</a></li>
                    <li>- CD8 Volcano Plots - <a href="/Candida_Migratory_TCR/Figures_Tables/supplemental_6.jpeg">supplemental_8.jpeg</a></li>
                    <li>- CD8 Cell Cycle Score - <a href="/Candida_Migratory_TCR/Figures_Tables/supplemental_7.jpeg">supplemental_9.jpeg</a></li>
                    <li>- Figure 7: CD8 Gene Ontology Analysis - <a href="/Candida_Migratory_TCR/Figures_Tables/figure_7.jpeg">figure_6.jpeg</a></li>

                    <li>- Figure 10: Neuro-Related Gene Expression - <a href="/Candida_Migratory_TCR/Figures_Tables/figure_10.jpeg">figure_10.jpeg</a></li>
                </ul>
            </li>
            <li>
                <a href="/Candida_Migratory_TCR/Scripts/R/sc_RNA_scripts/Candida_sc_clust5_analysis.Rmd">Candida_sc_clust5_analysis.Rmd</a>
                - Sub-clustering analysis for particular cluster of interest in scRNA-seq data.
                <ul>
                    <li><a href="/Candida_Migratory_TCR/Scripts/R/sc_RNA_scripts/WNN_script/Candida_sc_Analysis_WNN.Rmd">Candida_sc_Analysis_WNN.Rmd</a>
                        - Weighted nearest neighbor analysis for integration of scRNA with Cite-Seq surface protein assay.</li>
                </ul>
            </li>
        </ul>
        <p>GitHub Repository: <a href="https://github.com/NateRMorris/Candida_Migratory_TCR/tree/main/sc_RNA_scripts" target="_blank">R Scripts Repository</a></p>
    </section>

    <section id="tcr-scripts">
        <h2>R Scripts for TCR Analysis and Integration</h2>
        <p>Scripts related to TCR analysis and diversity.</p>
        <ul>
            <li><a href="/Candida_Migratory_TCR/Scripts/R/tcr_Scripts/Candida_tcr_paired.Rmd">Candida_tcr_paired.Rmd</a>
                - Indentifying and analyzing TCR sequences that are identical and found in cells from multiple organs, also considered "paired across organ", or migratory between organ.</li>
            <li><a href="/Candida_Migratory_TCR/Scripts/R/tcr_Scripts/Candida_tcr_duplicate.Rmd">Candida_tcr_duplicate.Rmd</a>
                - Identification and analyzing "duplicate" TCR sequences, which are found in multiple cells and therefore are proliferating.</li>
            <li><a href="/Candida_Migratory_TCR/Scripts/R/tcr_Scripts/Candida_tcr_seurat_integration.Rmd">Candida_tcr_seurat_integration.Rmd</a>
                - Integration of TCR data with Seurat scRNA data.</li>
            <li><a href="/Candida_Migratory_TCR/Scripts/R/tcr_Scripts/Candida_tcr_Diversity.Rmd">Candida_tcr_Diversity.Rmd</a>
                - Diversity analysis of TCR sequences.</li>
            <li><a href="/Candida_Migratory_TCR/Scripts/R/tcr_Scripts/Candida_tcr_clust5.Rmd">Candida_tcr_clust5.Rmd</a>
                - TCR analysis of particular cluster of interest.</li>
            <li><a href="/Candida_Migratory_TCR/Scripts/R/tcr_Scripts/Candida_tcr_Gearty_assumption.Rmd">Candida_tcr_Gearty_assumption.Rmd</a>
                - Analysis based on Gearty's assumptions.</li>
            <li><a href="/Candida_Migratory_TCR/Scripts/R/tcr_Scripts/Candida_tcr_Cytoscape.Rmd">Candida_tcr_Cytoscape.Rmd</a>
                - Visualization using Cytoscape.</li>
        </ul>
        <p>GitHub Repository: <a href="https://github.com/NateRMorris/Candida_Migratory_TCR/tree/main/tcr_Scripts" target="_blank">TCR Scripts Repository</a></p>
    </section>

    <section id="python-scripts">
        <h2>Python Scripts</h2>
        <p>Scripts related to TCR analysis using Python.</p>
        <ul>
            <li><a href="/Candida_Migratory_TCR/Scripts/R/py_scripts/TCR_Scipy.ipynb">TCR_Scipy.ipynb</a>
                - Using scipy for TCR analysis.</li>
        </ul>
        <p>GitHub Repository: <a href="https://github.com/NateRMorris/Candida_Migratory_TCR/tree/main/py_scripts" target="_blank">TCR Scripts Repository</a></p>
    </section>

    <section id="tcr_Packages">
        <h2>TCR Packages</h2>
        <p>Package information used in analysis.</p>
        <ul>
            <li><a href="/Candida_Migratory_TCR/blob/main/tcr_Packages/migratory_tcr.yml">TCR_Scipy.ipynb</a>
                - Environment file to reconstruct dependencies necessary for python analysis.</li>
            <li><a href="/Candida_Migratory_TCR/tree/main/tcr_Packages/Cytoscape%20Fig">TCR_Scipy.ipynb</a>
                - Cytoscape figure and file to remake figure.</li>
        </ul>
        <p>GitHub Repository: <a href="https://github.com/NateRMorris/Candida_Migratory_TCR/tree/main/tcr_Packages" target="_blank">TCR Scripts Repository</a></p>
    </section>

    <section id="Figures_Tables">
        <h2>Figures/Tables</h2>
        <p>Figures and Tables created during analysis, including Cytoscape Plots and Differentially Expressed Genes/Surface proteins per cluster and organ</p>

        <p>GitHub Repository: <a href="https://github.com/NateRMorris/Candida_Migratory_TCR/tree/main/tcr_Packages" target="_blank">TCR Scripts Repository</a></p>
    </section>
