---
title: "Candida_sc_Analysis_WNN"
author: "Nathan Morris"
date: "2023-11-16"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

### Data Import
```{r echo = TRUE, include= TRUE, error = FALSE, warning = FALSE, message=FALSE}
# Data Import
library(Seurat)
library(scater)
library(dplyr)
library(SingleR)
library(celldex)
library(scRNAseq)
library(scran)
library(gridExtra)
library(pheatmap)
library(ess)

data_dir <- "/Users/johnmorris/Desktop/Comp_Bio_MS_Weill_Cornell/Summer_2023/Betel Lab/Betel:Iliev Candida Project/Candida_Migratory_TCR_Project/Data"
setwd(data_dir)
```

Load the file from directory with annotations, so re-run isn't necessary
```{r}
seurat_object <- readRDS(paste0(data_dir, "dataset_post_sc_Analysis.rds"))
```
```
variable_genes <- VariableFeatures(seurat_object)
subset_data <- seurat_object[variable_genes,]
```


## Hartigans Dip Test

```
library(diptest)

gene_expression <- GetAssayData(seurat_object, assay = "RNA")
dip_result <- dip(gene_expression)
dip_statistic <- dip_result$D
p_value <- dip_result$p.value
# If the p-value is below the threshold, you can conclude that there is evidence of multimodality.

```
The Hartigan's dip test measures multimodality in a sample by the maximumm difference across all sample points, between the empirical distribution function and the unimodal function that minimizes that maximum difference.

Multimodality allows us to detect the subpopulations within the assay, while empirical distribution function is a form of cumulative distribution function, and is the sum of all points equal to or lower in expression.

Since we expect the data to follow a Gaussian/normal distribution for each individual population, we can expect there to be significant deviation from unimodality. This could allow us to detect if there are multimodality (subpopulations) within the assay data, and therefore check the quality of our assay. 


# WNN using PCA from RNA and PCA from ADT

Weighted nearest neighbor analysis allows us to integrate multi-modal analysis into one reduced dimensional space. It works first calculating sets of nearest neighbors using each modality. Then independently averaging the molecular profiles of protein neighbors and RNA neighbors (i.e. by predicting the molecular contents of a cell from it's neighbors) and comparing the averages to their original measured values.

Therefore, there are three key steps: (1) obtaining within modality and cross-modality predictions, (2) converting these to prediction affinities, based on a cell-specific bandwidth kernel, and (3) calculating modality weights using a softmax transformation. 

https://www.cell.com/cell/fulltext/S0092-8674(21)00583-3

```{r}
DefaultAssay(seurat_object) <- 'RNA'
seurat_object <- NormalizeData(seurat_object) %>% FindVariableFeatures() %>% ScaleData() %>% RunPCA()

DefaultAssay(seurat_object) <- 'ADT'
#VariableFeatures(seurat_object) <- rownames(seurat_object[["ADT"]])
seurat_object <- NormalizeData(seurat_object, normalization.method = 'CLR', margin = 2) %>% 
  ScaleData() %>% RunPCA(reduction.name = 'apca')
```

Based on the elbow plot, which shows how much variability / information is accounted for in each Principle Component, I've decided to use 14 columns from the PCA and 10 APCA based on the Elbow plot.
```{r}
ElbowPlot(seurat_object, reduction = "pca")
ElbowPlot(seurat_object, reduction = "apca")
```

```{r}
seurat_object <- FindMultiModalNeighbors(
  seurat_object, reduction.list = list("pca", "apca"), 
  dims.list = list(1:14, 1:10), modality.weight.name = c("RNA.weight", "ADT.weight")) 
# If I wanted to use all dimensions, it would be list(1:30, 1:18)

seurat_object <- RunUMAP(seurat_object, nn.name = "weighted.nn", reduction.name = "wnn.umap", reduction.key = "wnnUMAP_")

# We are going to Find Clusters using resolution 0.6 for the initial analysis and differential abundance
seurat_object <- FindClusters(seurat_object, graph.name = "wsnn", algorithm = 3, resolution = 0.6, verbose = FALSE)
```

Here, we could choose to just the wsnn or the wknn for making the clusters. This is the choice between the weighted shared nearest neighbors vs the weighted k-nearest neighbors. Shared is more appropriate to deal with noise and identify global structure, while k-nn is more appropriate when trying to identify local structure. Shared will make more dense and interconnected clusters, while k-nn will yield sparser clusters with more distinct populations.

I test with the Weighted Nearest Neighbor resolution being 0.6. This seems to create distinct clusters but doesn't over-cluster.
```{r}
p1 <- DimPlot(seurat_object, reduction = 'wnn.umap', group.by = "wsnn_res.0.6", label = TRUE, repel = TRUE, label.size = 2.5) + NoLegend()
p2 <- DimPlot(seurat_object, reduction = 'wnn.umap', group.by = 'RNA_clusters', label = TRUE, repel = TRUE, label.size = 2.5)
p1 + p2

Idents(seurat_object) <- 'wsnn_res.0.6'
seurat_object[['WNN_clusters_PCA']] <- Idents(seurat_object)
```

This shows the RNA/ADT weight used when creating the clusters.
```{r}
VlnPlot(seurat_object, features = "RNA.weight", group.by = 'WNN_clusters_PCA', sort = TRUE, pt.size = 0.1) + NoLegend()
```

Here are how the clusters are defined when using the 0.6 resolution.
```{r}
DimPlot(seurat_object, reduction = 'wnn.umap', group.by = 'WNN_clusters_PCA', label = TRUE, repel = TRUE, label.size = 2.5) + ggtitle("WNN UMAP by WNN Clusters")
```

Organ separates the clusters well. But this could be due to a certain organ having a higher proportion of a T-cell state than other organs. 
```{r}
DimPlot(seurat_object, reduction = 'wnn.umap', group.by = "Organ", label = FALSE, repel = TRUE, label.size = 2.5) + ggtitle("WNN UMAP by Organ")
```

Infection doesn't separate the clusters very well. It seems most clusters are present in both infected and non-infected mice.
```{r}
DimPlot(seurat_object, reduction = 'wnn.umap', group.by = 'Infection', label = TRUE, repel = TRUE, label.size = 2.5)
DimPlot(seurat_object, reduction = "wnn.umap", label = FALSE, repel = TRUE, split.by = "Infection") +  ggtitle("WNN UMAP by Infection")
```

It also doesn't separate very well by Mouse. Which is good, since we want replicates to be similar.
```{r}
Idents(object = seurat_object) <- "Mouse"
DimPlot(seurat_object, reduction = "wnn.umap", label = TRUE, label.size = 2, repel = TRUE) + NoLegend() +  ggtitle("WNN UMAP by Mouse")
```

You can see that the WNN clusters align with RNA clusters well.
```{r}
DimPlot(seurat_object, reduction = 'wnn.umap', group.by = "RNA_clusters", label = TRUE, repel = TRUE, label.size = 2.5) +  ggtitle("WNN UMAP by RNA Clusters")
```

WNN UMAP by SingleR annotations
```
# By Main RNA Annotation
Idents(object = seurat_object) <- "Immgen.main.RNA"
DimPlot(seurat_object, reduction = "wnn.umap", label = TRUE, repel = TRUE) +  ggtitle("WNN UMAP with RNA main Annotations")

# By Fine RNA Annotation
Idents(object = seurat_object) <- "Immgen.fine.RNA"
DimPlot(seurat_object, reduction = "wnn.umap", label = TRUE, label.size = 2, repel = TRUE) + NoLegend() +  ggtitle("WNN UMAP with RNA fine Annotations")
```

## ProjecTILS for T-cell state annotation
```{r}
library(ProjecTILs)
# https://github.com/carmonalab/ProjecTILs/tree/master
```

ProjecTILs is a computational method to project scRNA-seq data into reference single-cell atlases, enabling their direct comparison in a stable, annotated system of coordinates. The above loads the reference TIL atlas/map. Then you have a query dataset, or a query gene matrix with gene names and barcodes, and then it batch-corrects the query dataset and projects it into a low-dimensional embedding compatible with the reference.

https://pubmed.ncbi.nlm.nih.gov/34017005/
The above paper talks about it's use to interpret T-cell states from single cell transcriptomics. 

TIL is a play on tumor-infiltrating lymphocytes. However, they also have CD4/8 acute and chronic viral infection reference atlases as well.

```{r}
ref <- load.reference.map("/Users/johnmorris/Desktop/Comp_Bio_MS_Weill_Cornell/Summer_2023/Betel Lab/Betel:Iliev Candida Project/Candida_Migratory_TCR_Project/tcr_Packages/ProjecTILs/ref_TILAtlas_mouse_v1.rds") # This default takes from # https://doi.org/10.6084/m9.figshare.12478571
DefaultAssay(object = seurat_object) <- "RNA"
seurat_object <- ProjecTILs.classifier(query = seurat_object, ref = ref)
# This reference is the moust TIL atlas from melanoma and adenocarcinoma tumors
# https://figshare.com/articles/dataset/ProjecTILs_murine_reference_atlas_of_tumor-infiltrating_T_cells_version_1/12478571

#This code creates a new metadata "functional.cluster". Here is our WNN UMAP with these annotations.
#seurat_object@meta.data[["functional.cluster"]]

Idents(seurat_object) <- "functional.cluster"
seurat_object[['TIL_State']] <- Idents(seurat_object)
```

```{r}
setwd("/Users/johnmorris/Desktop/Comp_Bio_MS_Weill_Cornell/Summer_2023/Betel Lab/Betel:Iliev Candida Project/Candida_Scripts/ProjecTILs/")
ref_CD4 <- load.reference.map("ref_LCMV_CD4_mouse_release_v1.rds")
seurat_object <- ProjecTILs.classifier(query = seurat_object, ref = ref_CD4)
# This reference is the mouse acute and chronic viral infection CD4 t-cell atlas
# https://doi.org/10.6084/m9.figshare.16592693

Idents(seurat_object) <- "functional.cluster"
seurat_object[['CD4_TIL_State']] <- Idents(seurat_object)
```

The above uses the ProjecTILs.classifier function, which would apply the reference projection, but not alter the current embeddings. This simply uses ProjecTILs as a cell type classifier, which annotates cell types in our seurat object without altering existing embeddings. Overall this reference atlas summarizes TIL diversity using nine broad cell subtypes with distinct phenotypes, functions, metabolic lifestyles, and preferential tissue distributions, and strongly supported by experimental evidence in murine models.

```{r}
DimPlot(seurat_object, reduction = 'wnn.umap', group.by = 'TIL_State', label = TRUE, repel = TRUE, label.size = 2.5) + ggtitle("WNN UMAP by ProjecTILs T-cell annotations")
```

If we instead wanted to use the reference's UMAP/projection, we could run this:
```
query.projected <- Run.ProjecTILs(seurat_object, ref = ref)
```

# FACS Sorting-Like Graph

```{r}
Idents(object = seurat_object) <- "WNN_Clusters.1.5"
DefaultAssay(seurat_object) <- 'ADT'

FACS_sort_graph <- function(marker1, marker2, cluster_of_interest) {
  facs_df <- as.data.frame(log2(seurat_object@assays[["ADT"]]@counts[marker1,] + 1))
  colnames(facs_df)[1] <- marker1
  facs_df[,2] <- log2(seurat_object@assays[["ADT"]]@counts[marker2,] + 1 )
  colnames(facs_df)[2] <- marker2
  facs_df[,3] <- seurat_object$WNN_Clusters.1.5
  colnames(facs_df)[3] <- "WNN_Cluster"

  ggplot(facs_df, aes(x = facs_df[,1], y = facs_df[,2], col = factor(WNN_Cluster == cluster_of_interest))) +
  # Plot all points with transparency (alpha) for better visualization of overlapping points
  geom_point(alpha = 0.1, show.legend = FALSE) +
  # Color the points where WNN_Cluster is 0 in red
  scale_color_manual(values = c("TRUE" = "red", "FALSE" = "black")) +
  # Customize the legend title
  labs(x = paste0(marker1, " level"), y = paste0(marker2, " level"), title = paste0(marker1, " vs ", marker2), col = paste0("Cluster = ", cluster_of_interest)) +
  theme_bw()
}

cluster_FACS_check <- 7

p1 <- FACS_sort_graph("CD8A", "CD4", cluster_FACS_check)
p2 <- FACS_sort_graph("CD62L", "CD44", cluster_FACS_check)

p1 + p2

# At least in cluster 7, we can see that they are CD4+ and activated (CD62L low and CD44 high)

#hist(seurat_object@assays[["ADT"]]@counts["CD8A",])
#hist(seurat_object@assays[["ADT"]]@counts["CD4",])
#ggplot(facs_df, aes(x = facs_df[,1], y = facs_df[,2])) + geom_point(col = rgb(0,0,0,0.1)) +
#  labs(x = paste0(marker1, " level"), y = paste0(marker2, " level"), title = paste0(marker1, " vs ", marker2)) + theme_bw()
```



# Visualizing the expression of canonical marker genes and proteins on the multimodal UMAP
```{r}
# How to search for a gene name
#seurat_object[["RNA"]][grepl("Tcf", rownames(seurat_object[["RNA"]])),][,1]
```

```{r}
DefaultAssay(seurat_object) <- 'RNA'

# Gene Groups from Dr. Iliev's Papers
gois <- c("Tcf7", "Slamf6", "Ccr7", "Cxcr6", "Foxp3", "Mki67") # from RNA assay
stem_like_gois <- c("Ccr7", "Il7r", "Sell", "Tcf7")
# "Smurf1"
activation_like_gois <- c("Cd44", "Cd69", "Il17a", "Il17f", "Il22", "Ifng")
ProjecTILs.markers <- c("Cd4", "Cd8a", "Foxp3", "Ccr7", "Slamf6", "Ifngr1", "Pdcd1", "Ctla4", "Gzmb", "Cx3cr1", "Tox")
```

Violin Plots of Specific Gene Groups

Using the ProjecTILs markers from their paper (https://www.nature.com/articles/s41467-021-23324-4), the marker genes in comparison to the groups seem pretty similar. Meaning that the annotation worked properly.
```{r}
DefaultAssay(seurat_object) <- 'RNA'
VlnPlot(seurat_object, ProjecTILs.markers, group.by = 'TIL_State')
```

```{r}
ADT_markers_of_interest <- list("CD8A", "CD4", "CD3", "CCR6", "IL21R", "SCA1") # from Cite-seq Assay

#https://www.rndsystems.com/resources/cell-markers/immune-cells/helper-t-cells/th17-cell-markers
# Cell surface markers (ADT) taken from here

Idents(object = seurat_object) <- "WNN_clusters_PCA"

# Expression Plots
FeaturePlot(seurat_object, features = c(activation_like_gois),
                  reduction = 'wnn.umap', max.cutoff = 2, 
                  cols = c("lightgrey","darkgreen"), ncol = 3)

FeaturePlot(seurat_object, features = c(ADT_markers_of_interest),
                  reduction = 'wnn.umap', max.cutoff = 2, 
                  cols = c("lightgrey","darkgreen"), ncol = 3)

#RidgePlot(seurat_object, features = c(stem_like_gois), ncol = 2)

#DotPlot(seurat_object, features = c(gois)) + RotatedAxis() + ggtitle("WNN Clusters by Th17 stem-like markers")

#DoHeatmap(seurat_object, features = c(gois)) + RotatedAxis() + ggtitle("WNN Clusters by Th17 stem-like markers")
```

Looking at specific T-cell type signatures from a previous paper.
```{r}
# https://www.nature.com/articles/s41590-020-00836-7#data-availability
# T-effector cell type signatures (Supplemental Data Table 2)
# However, this shows that cytokines do not represent main axes of variance on colonic T-cells
th1 <- c("Ifng", "Tbx21", "Cxcr3", "Nkg7", "Fasl", "Il2", "Tnf", "Il12rb1", "Il12rb2", "Gzmb")
th2 <- c("Gata3", "Il4", "Il5", "Il13", "Il10", "Areg", "Maf", "Il17rb")
th17 <- c("Rorc", "Il17a", "Il17f", "Il22", "Ccr6", "Il27r", "Tmem176a")
tfh <- c("Cxcr5", "Izumo1r", "Nt5e", "Bcl6", "Slamf6", "Cd200", "Ccr7", "Tcf7")

Idents(object = seurat_object) <- "WNN_clusters_PCA"

FeaturePlot(seurat_object, features = th17,
                  reduction = 'wnn.umap', max.cutoff = 2, 
                  cols = c("lightgrey","darkgreen"), ncol = 3)

DotPlot(seurat_object, features = th17) + RotatedAxis() + ggtitle("WNN Clusters by Th17-like markers")

DoHeatmap(seurat_object, features = c(th17)) + RotatedAxis() + ggtitle("WNN Clusters by Th17-like markers")

#VlnPlot(seurat_object, features = th17, group.by = 'WNN_clusters_PCA')
```

I would say these show that cluster 8 and 13 are the primary clusters containing Th17 cells based on this data. However, the paper that these gene lists were taken from, shows that the cytokines T-cells produce do not distinctly separate different T-cell subtypes. This can be seen from the other cell types (th1, th2, tfh), which don't really aggregate to a particular cluster. But also, these show that cluster 8 doesn't include many th1, th2 or tfh cell types either, which is complementing our conclusion.


# DimPlot by Organ
```{r}
organs <- c("Colon", "Lung", "mLns")

seurat_object.colon <- subset(x = seurat_object, subset = (Organ == "Colon"))
seurat_object.lung <- subset(x = seurat_object, subset = (Organ == "Lung"))
seurat_object.mlns <- subset(x = seurat_object, subset = (Organ == "mLns"))
```

I don't think I need to re-run FindMultiModalNeighbors again? I am just re-finding clusters that I have already determined from the whole seurat_object.
```
# Colon
seurat_object.colon <- FindMultiModalNeighbors(
  seurat_object.colon, reduction.list = list("pca", "apca"), 
  dims.list = list(1:14, 1:10), modality.weight.name = c("RNA.weight", "ADT.weight")
)
seurat_object.colon <- RunUMAP(seurat_object.colon, nn.name = "weighted.nn", reduction.name = "wnn.umap", reduction.key = "wnnUMAP_")
seurat_object.colon <- FindClusters(seurat_object.colon, graph.name = "wsnn", algorithm = 3, resolution = 0.6, verbose = FALSE)
unique(seurat_object.colon@meta.data[["wsnn_res.1"]]) # Using resolution = 1 --> Has 10 total clusters

# Lung
seurat_object.lung <- FindMultiModalNeighbors(
  seurat_object.lung, reduction.list = list("pca", "apca"), 
  dims.list = list(1:14, 1:10), modality.weight.name = c("RNA.weight", "ADT.weight")
)
seurat_object.lung <- RunUMAP(seurat_object.lung, nn.name = "weighted.nn", reduction.name = "wnn.umap", reduction.key = "wnnUMAP_")
seurat_object.lung <- FindClusters(seurat_object.lung, graph.name = "wsnn", algorithm = 3, resolution = 0.6, verbose = FALSE)
unique(seurat_object.lung@meta.data[["wsnn_res.1"]]) # Using resolution = 1 --> Has 10 total clusters

# mLns
seurat_object.mlns <- FindMultiModalNeighbors(
  seurat_object.mlns, reduction.list = list("pca", "apca"), 
  dims.list = list(1:14, 1:10), modality.weight.name = c("RNA.weight", "ADT.weight")
)
seurat_object.mlns <- RunUMAP(seurat_object.mlns, nn.name = "weighted.nn", reduction.name = "wnn.umap", reduction.key = "wnnUMAP_")
seurat_object.mlns <- FindClusters(seurat_object.mlns, graph.name = "wsnn", algorithm = 3, resolution = 0.6, verbose = FALSE)
unique(seurat_object.mlns@meta.data[["wsnn_res.1"]]) # Using resolution = 1 --> Has 10 total clusters
```

```{r}
Idents(object = seurat_object.colon) <- "WNN_clusters_PCA"
DimPlot(seurat_object.colon, reduction = "wnn.umap", label = TRUE, repel = TRUE) +  ggtitle("Colon WNN UMAP by Cluster")
#WNN.colon <- FindAllMarkers(seurat_object.colon, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)

#Idents(object = seurat_object.colon) <- "Immgen.fine.RNA"
Idents(object = seurat_object.colon) <- "Infection"
DimPlot(seurat_object.colon, reduction = "wnn.umap", label = TRUE, repel = TRUE) +  ggtitle("Colon WNN UMAP by Infection")

Idents(object = seurat_object.lung) <- "WNN_clusters_PCA"
DimPlot(seurat_object.lung, reduction = "wnn.umap", label = TRUE, repel = TRUE) +  ggtitle("Lung WNN UMAP by Cluster")
#WNN.lung <- FindAllMarkers(seurat_object.lung, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)

#Idents(object = seurat_object.lung) <- "Immgen.fine.RNA"
Idents(object = seurat_object.lung) <- "Infection"
DimPlot(seurat_object.lung, reduction = "wnn.umap", label = TRUE, repel = TRUE) +  ggtitle("Lung WNN UMAP by Infection")

Idents(object = seurat_object.mlns) <- "WNN_clusters_PCA"
DimPlot(seurat_object.mlns, reduction = "wnn.umap", label = TRUE, repel = TRUE) +  ggtitle("mLns WNN UMAP by Cluster")
#WNN.mlns <- FindAllMarkers(seurat_object.mlns, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)

#Idents(object = seurat_object.mlns) <- "Immgen.fine.RNA"
Idents(object = seurat_object.mlns) <- "Infection"
DimPlot(seurat_object.mlns, reduction = "wnn.umap", label = TRUE, repel = TRUE) +  ggtitle("mLns WNN UMAP by Infection")
```


#Visualization of differentially abundant clusters for Weighted Nearest Neighbors
```{r}
#library(reshape2)
#df.long <- melt(df, id.vars=c("Organ", "Cluster"))
library(tidyr)

df$Organ_Cluster <- paste(df$Organ, df$Cluster)

df %>%
    pivot_longer(CA_Proportion:NT_Proportion, names_to = "Infection", values_to = "Proportion") %>%
    ggplot(aes(Organ_Cluster, Proportion, fill = Infection)) +
    geom_col(position = "dodge") +
    geom_text(aes(label = ifelse(adj_p_value < 0.05, "*", "")), 
              position = position_dodge(width = 0.9), vjust = -0.5, size = 4) +
    theme(axis.text.x = element_text(angle = 45, hjust = 1))
```

DF for each Organ
```{r}
organs <- c("Colon", "Lung", "mLns")

for (organ in organs){
    df.subset <- df[df$Organ == organ,]
    
    plot <- df.subset %>%
    pivot_longer(CA_Proportion:NT_Proportion, names_to = "Infection", values_to = "Proportion") %>%
    ggplot(aes(Cluster, Proportion, fill = Infection)) +
    geom_col(position = "dodge") +
    geom_text(aes(label = ifelse(adj_p_value < 0.05, "*", "")), 
              position = position_dodge(width = 0.9), vjust = -0.5, size = 4) +
    theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
    ggtitle(paste0(organ, " Abundance Difference"))
    print(plot)
}
```

Finding Marker Genes and Making Heat Map for Cluster 7:
```{r}
Idents(object = seurat_object) <- "WNN_clusters_PCA_1.3"
clust7.markers <- FindMarkers(seurat_object, ident.1 = 7, only.pos = TRUE)

markers_7 <- clust7.markers[order(clust7.markers$p_val_adj), ]
markers_7 <- markers_7[(markers_7$p_val_adj > 0),]

marker_genes_7 <- rownames(markers_7[1:8,])
marker_genes_7 <- marker_genes_7[!is.na(marker_genes_7)]

more_marker_genes_7 <- rownames(markers_7[1:50,])
DoHeatmap(seurat_object, features = more_marker_genes_7, raster = FALSE)

```

#Volcano Plot for Cluster 7
```{r}
Idents(object = seurat_object) <- "WNN_clusters_PCA_1.3"
clust7_all.markers <- FindMarkers(seurat_object, ident.1 = 7)
clust7_all.markers <- rownames_to_column(clust7_all.markers,"gene")
```

```{r}
outlier_max_pval = 1.0e-100
outlier_max_logFC = 3
outlier_min_logFC = -3

DEGs <- clust7_all.markers %>%
  mutate(p_val= ifelse(p_val < outlier_max_pval, outlier_max_pval, p_val)) %>%
  mutate(avg_log2FC = ifelse(avg_log2FC < outlier_min_logFC, outlier_min_logFC, avg_log2FC)) %>%
  mutate(avg_log2FC = ifelse(avg_log2FC > outlier_max_logFC, outlier_max_logFC, avg_log2FC))

EnhancedVolcano(DEGs,
  lab= DEGs$gene,
  x='avg_log2FC',
  y= 'p_val',
  xlim = c(-3,3),
  ylim = c(0,100),
  title = paste0('RNA-clusters: 7 vs all'),
  pCutoff = 0.05,
  FCcutoff = 0.15,
  labSize = 1)
```

#Finding Markers Genes and Making Heat Map for Cluster 14:
```{r}
Idents(object = seurat_object) <- "WNN_clusters_PCA_1.3"
clust14.markers <- FindMarkers(seurat_object, ident.1 = 14, only.pos = TRUE)

markers_14 <- clust14.markers[order(clust14.markers$p_val_adj), ]
markers_14 <- markers_14[(markers_14$p_val_adj > 0),]

marker_genes_14 <- rownames(markers_14[1:8,])
marker_genes_14 <- marker_genes_14[!is.na(marker_genes_14)]

more_marker_genes_14 <- rownames(markers_14[1:50,])
DoHeatmap(seurat_object, features = more_marker_genes_14, raster = FALSE)
```


# Proptest - Candida infected cell # vs Non-infected # for all organs and WNN_PCA clusters
```{r, warning = FALSE}
cluster_differences <- function(cluster_name){
  
  organs <- c("Colon", "Lung", "mLns")
  
  df_organ <- c()
  df_cluster <- c()
  df_pcNT <- c()
  df_pcCA <- c()
  df_pval <- c()
  
  n <- 1
  
  WNN_organ <- c(seurat_object.colon, seurat_object.lung, seurat_object.mlns)
  cluster_name <- as.character(cluster_name)
  
  for (seurat_object.sub in WNN_organ) {
    
    organ <- as.character(unique(seurat_object.sub$Organ))
    clusters = sort(unique(seurat_object.sub[[cluster_name]][[1]]))
    
    for (clust in clusters) {
  
      NTc <- nrow(seurat_object.sub@meta.data[seurat_object.sub[[cluster_name]][[1]] == clust & seurat_object.sub@meta.data$Infection == "No", ])
      # NTc is the number of no Candida from an organ in a particular individual cluster
      CAc <- nrow(seurat_object.sub@meta.data[seurat_object.sub[[cluster_name]][[1]] == clust & seurat_object.sub@meta.data$Infection == "Ca", ])
      # CAc is the number of Candida from an organ in a particular individual cluster
      
      NT <- nrow(seurat_object.sub@meta.data[seurat_object.sub@meta.data$Infection == "No", ])
      # Total amount of No infection from the organ
      CA <- nrow(seurat_object.sub@meta.data[seurat_object.sub@meta.data$Infection == "Ca", ])
      # Total amount of Ca infection from the organ
      
      successes <- c(NTc, CAc)
      trials <- c(NT, CA)
      
      result <- tryCatch({
    res <- prop.test(x = successes, n = trials)
      }, warning = function(w) {
    # Code to handle warnings
    warningCondition <- conditionMessage(w)
    cat("Warning:", warningCondition, "\n")
      })
      # Prop test to test the differences between the number of No/Ca in that cluster
      # Versus the total number of those samples from the organ
      
      df_organ[n] <- organ
      df_cluster[n] <- clust
      df_pcNT[n] <- NTc/NT
      df_pcCA[n] <- CAc/CA
      df_pval[n] <- res$p.value
      
      n <- n + 1
    }
  }  
  df <- data.frame(Organ = df_organ, Cluster = df_cluster, NT_Proportion = df_pcNT, CA_Proportion = df_pcCA, p_value = df_pval)
  df$adj_p_value <- p.adjust(df$p_value, method = "bonferroni")
  return(df)
}
```

```{r}
seurat_object <- FindClusters(seurat_object, graph.name = "wsnn", algorithm = 1, resolution = 0.6, verbose = FALSE)
seurat_object <- FindClusters(seurat_object, graph.name = "wsnn", algorithm = 3, resolution = 0.4, verbose = FALSE)
seurat_object <- FindClusters(seurat_object, graph.name = "wsnn", algorithm = 2, resolution = 0.3, verbose = FALSE)
seurat_object.colon <- subset(x = seurat_object, subset = (Organ == "Colon"))
seurat_object.lung <- subset(x = seurat_object, subset = (Organ == "Lung"))
seurat_object.mlns <- subset(x = seurat_object, subset = (Organ == "mLns"))
#Idents(seurat_object.colon) <- 'WNN_clusters_PCA'
#Idents(seurat_object.mlns) <- 'WNN_clusters_PCA'
#Idents(seurat_object.lung) <- 'WNN_clusters_PCA'
df <- cluster_differences('wsnn_res.0.6')
df <- cluster_differences('RNA_clusters')
df <- cluster_differences('wsnn_res.0.4')
df <- cluster_differences('wsnn_res.0.3')
```

```{r}
df <- cluster_differences('wsnn_res.0.5')
df[df$adj_p_value < 0.05,]
table(seurat_object$TIL_State[seurat_object$wsnn_res.0.4 == 4])
table(seurat_object$wsnn_res.0.6)
```


```{r}
#library(reshape2)
#df.long <- melt(df, id.vars=c("Organ", "Cluster"))
library(tidyr)
df.long <- df %>%
  pivot_longer(cols = c(NT_Proportion, CA_Proportion),
               names_to = "Proportion_Type",
               values_to = "Proportion_Value")

ggplot(df.long, aes(x = Cluster, y = Proportion_Value, fill = Organ)) +
  geom_bar(stat = "identity", position = "dodge") +
  facet_wrap(~Proportion_Type, scales = "free_y") +
  labs(x = "Cluster", y = "Proportion") +
  theme_minimal()
```

# Differential Abundance using DESeq2

```{r}
library(DESeq2)


```



# Differential Abundance using Milo
https://bioconductor.org/books/3.12/OSCA/multi-sample-comparisons.html#differential-abundance

```{r}
#BiocManager::install("miloR")
#https://bioconductor.org/packages/release/bioc/vignettes/miloR/inst/doc/milo_demo.html

# BUMMER BELOW
#In progress: we are perfecting the functionality to add a precomputed KNN graph (for example constructed with Seurat or scanpy) to the graph slot using the adjacency matrix.
#https://bioconductor.org/packages/release/bioc/vignettes/miloR/inst/doc/milo_demo.html

library(SingleCellExperiment)
library(miloR)

#table(seurat_object$functional.cluster, merged$sample) 

sce <- as.SingleCellExperiment(seurat_object)
reducedDim(sce, "wnn.UMAP") <- seurat_object@reductions[["wnn.umap"]]@cell.embeddings

plotReducedDim(sce, colour_by="functional.cluster", dimred = "UMAP_RNA")
plotReducedDim(sce, colour_by="functional.cluster", dimred = "wnn.UMAP")


table(seurat_object$WNN_clusters_PCA[seurat_object$Organ == "Colon"], seurat_object$Infection[seurat_object$Organ == "Colon"])

nrow(seurat_object.colon@meta.data[seurat_object.colon@meta.data$Infection == "No", ])
sum(seurat_object$Organ == "Colon" & seurat_object$Infection  == "No")
sum(seurat_object$Infection  == "Ca")

sce_milo <- Milo(sce)
sce_milo <- buildGraph(sce_milo, reduced.dim = "wnn.UMAP")
#sce_milo <- countCells(sce_milo, meta.data = data.frame(colData(sce_milo)), samples="sample_name")
sce_milo <- countCells(sce_milo, meta.data = data.frame(colData(sce_milo)), samples="Infection")


```



# Cluster Similarity

Here, I check the similarity between the RNA and WNN clusters

These scales are represented as the Intersect of Cluster A and B (common) / the Union of Cluster A and B (difference)

```{r}
compare_cluster1 <- seurat_object.colon$wsnn_res.0.4
compare_cluster2 <- seurat_object.colon$RNA_clusters

cluster_intersection <- matrix(nrow = length(unique(compare_cluster1)), ncol = length(unique(compare_cluster2)))
cluster_union <- matrix(nrow = length(unique(compare_cluster1)), ncol = length(unique(compare_cluster2)))

for (i in seq(1:nrow(cluster_intersection))){
  cluster.i <- i - 1
  numi <- compare_cluster1 == cluster.i
  for (j in seq(1:ncol(cluster_intersection))){
    cluster.j <- j - 1
    numj <- compare_cluster2 == cluster.j
    cluster_intersection[i,j] <- sum(numi == TRUE & numj == TRUE)
    cluster_union[i,j] <- sum(numi == TRUE | numj == TRUE)
  }
}

cluster_matrix <- cluster_intersection / cluster_union
rownames(cluster_matrix) <- paste("WNN_PCA", sort(unique(compare_cluster1)))
colnames(cluster_matrix) <- paste("RNA", sort(unique(compare_cluster2)))

# Create heatmap
pheatmap(cluster_matrix, 
        #col = colorRampPalette(c("yellow", "blue"))(20),
        main = "Cluster Similarity Heatmap between WNN_PCA and RNA")
        #labels_row = make_bold_names(cluster_matrix, rownames, c("WNN 4")),
        #labels_col = make_bold_names(cluster_matrix, colnames, c("RNA 5")))
        #xlab = "RNA Clusters",
        #ylab = "WNN Clusters")
```

```{r}
compare_cluster1 <- seurat_object$Protein_clusters
compare_cluster2 <- seurat_object$RNA_clusters

cluster_intersection <- matrix(nrow = length(unique(compare_cluster1)), ncol = length(unique(compare_cluster2)))
cluster_union <- matrix(nrow = length(unique(compare_cluster1)), ncol = length(unique(compare_cluster2)))

for (i in seq(1:nrow(cluster_intersection))){
  cluster.i <- i - 1
  numi <- compare_cluster1 == cluster.i
  for (j in seq(1:ncol(cluster_intersection))){
    cluster.j <- j - 1
    numj <- compare_cluster2 == cluster.j
    cluster_intersection[i,j] <- sum(numi == TRUE & numj == TRUE)
    cluster_union[i,j] <- sum(numi == TRUE | numj == TRUE)
  }
}

cluster_matrix <- cluster_intersection / cluster_union
rownames(cluster_matrix) <- paste("ADT", sort(unique(compare_cluster1)))
colnames(cluster_matrix) <- paste("RNA", sort(unique(compare_cluster2)))

new_matrix <- cbind(cluster_matrix[,"RNA 3"], cluster_matrix[,"RNA 2"], cluster_matrix[,"RNA 1"], cluster_matrix[,"RNA 0"], cluster_matrix[,"RNA 4"], cluster_matrix[,"RNA 5"], cluster_matrix[,"RNA 6"], cluster_matrix[,"RNA 10"], cluster_matrix[,"RNA 7"], cluster_matrix[,"RNA 11"], cluster_matrix[,"RNA 12"], cluster_matrix[,"RNA 9"], cluster_matrix[,"RNA 13"], cluster_matrix[,"RNA 14"], cluster_matrix[,"RNA 15"],cluster_matrix[,"RNA 16"] )
colnames(new_matrix) <- c("RNA 3","RNA 2","RNA 1","RNA 0","RNA 4","RNA 5","RNA 6","RNA 10","RNA 7","RNA 11", "RNA 12", "RNA 9","RNA 13","RNA 14","RNA 15","RNA 16")

pheatmap(new_matrix,main = "Cluster Similarity Heatmap between RNA and ADT", cluster_rows = FALSE, cluster_cols = FALSE)

```

# Monocle3
```{r}
#devtools::install_github("satijalab/seurat-wrappers")
library(monocle3)
```

```{r}
DefaultAssay(seurat_object) <- "RNA"

seurat_object.colon <- subset(x = seurat_object, subset = (Organ == "Colon"))
seurat_object.lung <- subset(x = seurat_object, subset = (Organ == "Lung"))
seurat_object.mlns <- subset(x = seurat_object, subset = (Organ == "mLns"))
```

Here, I'm dividing the clusters with a higher resolution. Here, we are hoping to get the cluster 
```{r}
seurat_object <- FindClusters(seurat_object.colon, graph.name = "wsnn", algorithm = 3, resolution = 0.6, verbose = FALSE)
seurat_object <- FindClusters(seurat_object.lung, graph.name = "wsnn", algorithm = 3, resolution = 0.6, verbose = FALSE)
seurat_object <- FindClusters(seurat_object.mlns, graph.name = "wsnn", algorithm = 3, resolution = 0.6, verbose = FALSE)
```

```{r}
cds <- SeuratWrappers::as.cell_data_set(seurat_object.colon)

cds
# colData(cds) # Metadata here
#fData(cds) # Gene Metadata (empty)
#rownames(fData(cds)) # Gene names
# counts(cds) # Gives counts
fData(cds)$gene_short_name <- rownames(fData(cds)) # Requires short names
#fData(cds)$gene_short_name <- rownames(cds) # Requires short names


recreate.partition <- c(rep(1,length(cds@colData@rownames)))
names(recreate.partition) <- cds@colData@rownames
recreate.partition <- as.factor(recreate.partition)

cds@clusters$UMAP$partitions <- recreate.partition

WNN_clusters <- seurat_object.colon$WNN_clusters_PCA

cds@clusters$UMAP$clusters <- WNN_clusters

# Assign UMAP coordinate - cell embeddings

cds@int_colData@listData$reducedDims$UMAP<- seurat_object.colon@reductions$wnn.umap@cell.embeddings


pData(cds)$sample_name <- NULL


cluster.before.trajectory <- plot_cells(cds, color_cells_by = "cluster", label_groups_by_cluster = FALSE, group_label_size = 5) +
  theme(legend.position = "right")

cds <- learn_graph(cds, use_partition = FALSE)

plot_cells(cds,
           color_cells_by = "cluster",
           label_groups_by_cluster = FALSE,
           label_branch_points = FALSE,
           label_roots = FALSE,
           label_leaves = FALSE,
           group_label_size = 5)
```

Ordering by pseudotime - not sure what to make the root cells
```{r}
cds <- order_cells(cds, reduction_method = "UMAP", root_cells = colnames(cds[,clusters(cds) == 0]))
# I would want to use wnn.UMAP, but gives an error
```

```{r}
plot_cells(cds,
           color_cells_by = "pseudotime",
           label_groups_by_cluster = FALSE,
           label_branch_points = FALSE,
           label_roots = FALSE,
           label_leaves = FALSE,
           group_label_size = 5)
```

```{r}
cds$monocle3_pseudotime <- pseudotime(cds)
data.pseudo <- as.data.frame(colData(cds))
#ggplot(data.pseudo, aes(monocle3_pseudotime, WNN_clusters_PCA, fill = WNN_clusters_PCA)) + geom_boxplot()
#ggplot(data.pseudo, aes(monocle3_pseudotime, Immgen.main.RNA, fill = Immgen.main.RNA)) + geom_boxplot()
ggplot(data.pseudo, aes(monocle3_pseudotime, wsnn_res.2, fill = wsnn_res.2)) + geom_boxplot()
```

# Diffusion Map
```
library(SingleCellExperiment)
library(scater)
library(TSCAN)
library(destiny)

seurat_object.clust5 <- subset(x = seurat_object, subset = (RNA_clusters == 5))

Idents(seurat_object.clust5) <- "Immgen.fine.RNA"
DefaultAssay(object = seurat_object.clust5) <- "RNA"

data_sce5 <- as.SingleCellExperiment(seurat_object.clust5)

# Add metadata
coldata <- DataFrame(
    genotype=seurat_object.clust5@meta.data$wsnn_res.1
    #phenotype=coldata[,"Characteristics[phenotype]"],
    #spike_in=coldata[,"Factor Value[spike-in addition]"],
    #row.names=coldata[,"Source Name"]
)
colData(data_sce5) <- coldata

table(data_sce5$genotype) # This has the cell identification based on the Seurat clusters

dm <- DiffusionMap(data_sce5, verbose = TRUE) 

cellLabels <- data_sce5$genotype
#sample_name <- data_sce$sample_name
tmp <- data.frame(DC1 = eigenvectors(dm)[, 1],
                  DC2 = eigenvectors(dm)[, 2],
                  DC3 = eigenvectors(dm)[, 3],
                  DC4 = eigenvectors(dm)[, 4],
                  # Clusters = cellLabels
                  # Mouse_Samples = sample_name
                  Timepoint = cellLabels) # Can use Timepoint or Samples here. Are they equivalent?
# Can also change this to sample_name, and see for each Mouse sample

ggplot(tmp, aes(x = DC1, y = DC2, colour = Timepoint)) + # Therefore, would want to change colour = Samples
  geom_point()  + 
  xlab("Diffusion component 1") + 
  ylab("Diffusion component 2") +
  theme_classic()
```

# Differential Expression 


Comparing All Clusters Infection vs Non-Infection
```{r}
num_of_cells_cluster <- matrix(nrow = 20, ncol = 3)
colnames(num_of_cells_cluster) <- c("Cluster", "Infected #", "Non-infected #")
infection_de_list_RNA <- list()
infection_de_list_ADT <- list()
for (cluster in unique(sort(seurat_object$WNN_Clusters.1.5))){
  print(cluster)
  cluster_seurat <- seurat_object[,seurat_object$WNN_Clusters.1.5 == cluster]

  num_of_cells_cluster[as.numeric(cluster) + 1, "Cluster"] <- cluster
  num_of_cells_cluster[as.numeric(cluster) + 1, "Infected #"] <- sum(cluster_seurat$Infection == "Ca")
  num_of_cells_cluster[as.numeric(cluster) + 1, "Non-infected #"] <- sum(cluster_seurat$Infection == "No")
  
  DefaultAssay(cluster_seurat) <- "RNA"
  Idents(object = cluster_seurat) <- "Infection"
  de_RNA <- FindMarkers(cluster_seurat, ident.1 = "Ca", only.pos = TRUE)
  #infection_de_list_RNA[[cluster]] <- de_RNA[de_RNA$p_val_adj < 0.05]
  if (nrow(de_RNA) > 0){
      infection_de_list_RNA[[cluster]] <- de_RNA[de_RNA$p_val_adj < 0.05,]
  } else {
      print(paste0("Cluster", cluster, "has no differential gene expression for infected cells"))
  }
  
  DefaultAssay(cluster_seurat) <- "ADT"
  Idents(object = cluster_seurat) <- "Infection"
  de_ADT <- FindMarkers(cluster_seurat, ident.1 = "Ca", only.pos = TRUE)
  if (nrow(de_ADT) > 0){
      infection_de_list_ADT[[cluster]] <- de_ADT[de_ADT$p_val_adj < 0.05,]
  } else {
      print(paste0("Cluster", cluster, "has no differential protein expression for infected cells"))
  }
}
```


```{r}
num_of_cells_cluster
infection_de_list_RNA
infection_de_list_ADT
```

As we can see here, there are very few differentially expressed genes between the infected and the non-infected cells within most clusters, however Cluster 8 in particular has many differentially expressed genes (Tmem176a, Ccr6, Ramp1). 

Cluster 3 shows Ifi27l2a, Cluster 2 shows Slc15a2 and Cd7, Cluster 1 shows Slc15a2 and Dnaja1 and finally Cluster 0 shows Slc15a2, Rflnb and Ifi27l2a.


Comparing All Clusters vs other CD8 or CD4 T cells
```{r}
cluster_de_list_CD4_RNA <- list()
cluster_de_list_CD4_ADT <- list()

CD4_clusters <- sort(c(0, 7, 9, 3, 5, 6, 8, 15, 17, 18, 16))
CD4_seurat <- seurat_object[,seurat_object$WNN_Clusters.1.5 == CD4_clusters]

for (cluster in CD4_clusters){
  print(cluster)
  DefaultAssay(CD4_seurat) <- "RNA"
  Idents(object = CD4_seurat) <- "WNN_Clusters.1.5"
  de_RNA <- FindMarkers(CD4_seurat, ident.1 = cluster, only.pos = TRUE)
  if (nrow(de_RNA) > 0){
      cluster_de_list_CD4_RNA[[as.character(cluster)]] <- de_RNA[de_RNA$p_val_adj < 0.05,]
  } else {
      print(paste0("Cluster", cluster, "has no differential gene expression for infected cells"))
  }
  
  DefaultAssay(CD4_seurat) <- "ADT"
  Idents(object = CD4_seurat) <- "WNN_Clusters.1.5"
  de_ADT <- FindMarkers(CD4_seurat, ident.1 = cluster, only.pos = TRUE)
  if (nrow(de_ADT) > 0){
      cluster_de_list_CD4_ADT[[as.character(cluster)]] <- de_ADT[de_ADT$p_val_adj < 0.05,]
  } else {
      print(paste0("Cluster", cluster, "has no differential protein expression for infected cells"))
  }
}

cluster_de_list_CD8_RNA <- list()
cluster_de_list_CD8_ADT <- list()

CD8_clusters <- sort(c(1, 14, 2, 10, 12, 19, 13, 11, 4))
CD8_seurat <- seurat_object[,seurat_object$WNN_Clusters.1.5 == CD8_clusters]

for (cluster in CD8_clusters){
  
  DefaultAssay(CD8_seurat) <- "RNA"
  Idents(object = CD8_seurat) <- "WNN_Clusters.1.5"
  de_RNA <- FindMarkers(CD8_seurat, ident.1 = cluster, only.pos = TRUE)
  cluster_de_list_CD8_RNA[[as.character(cluster)]] <- de_RNA[de_RNA$p_val_adj < 0.05,]
  
  DefaultAssay(CD8_seurat) <- "ADT"
  Idents(object = CD8_seurat) <- "WNN_Clusters.1.5"
  de_ADT <- FindMarkers(CD8_seurat, ident.1 = cluster, only.pos = TRUE)
  cluster_de_list_CD8_ADT[[as.character(cluster)]] <- de_ADT[de_ADT$p_val_adj < 0.05,]
}

```

By comparing all clusters vs the other CD4 or CD8 clusters, we are able to annotate each cluster with higher quality.

We can see that 

CD4 T cells:

Cluster 0 - Naive - Tcf7 and Ccr7
Cluster 7 - Naive , CD45RB
Cluster 9 - Naive -  Jun - could promote Th17 differentiation (https://www.nature.com/articles/s41467-017-00380-3/)
                      - early induction and sustained levels suggests it plays a role in Th17 differentiation
                    - CD2 HIGH

Cluster 3 - Modulating Th17 differentiation - high Ddit4 - (https://pubmed.ncbi.nlm.nih.gov/29378913/)
Cluster 5 - Modulating Th17 differentiation - high Ddit4 - (https://pubmed.ncbi.nlm.nih.gov/29378913/)
Cluster 18 - 

Cluster 6 - Treg - Foxp3, FR4+
Cluster 15 - Seems like it could be a mix of the Th17 and the Treg - also has a portion that is separate from the group (maybe recluster?), FR4 and CD73 HIGH - could be some sort of TAN (tumor associated neutrophil) - https://www.nature.com/articles/s41385-020-00349-4

Cluster 16 - Treg  - high Foxp3, Ikzf2 but also Gzmb , CD44 High, SlAM.CD150 High
CLuster 8 - Th17 Cells, CD44 HIGH
Cluster 17 - TCR gamma delta - CD44 high, CXCR6 HIGH
        -  the TCRGD is the most differentiated between cluster 17 and cluster 8 in ADT
```
DefaultAssay(seurat_object) <- 'ADT'
temp <- FindMarkers(seurat_object, ident.1 = 17,  ident.2 = 8, only.pos = TRUE)
```

CD8 T cells:

```{r}
DefaultAssay(seurat_object) <- "RNA"
Idents(object = seurat_object) <- "WNN_Clusters.1.5"

poo <- c("Prf1", "Gzmb", "Slamf7", "Plek", "Nkg7", "Crtam", "Ccl3", "Ccl4", "Ccl5", "Hopx", "Ifng", "Tbx21", "Csf2", "Il2", "Il22", "Il17A", "CCR7", "Sell", "CD27", "ICOS", "PDCD1", "CXCR5", "POU2AF1", "IL21" ,"CTLA4", "FOXP3", "LAG3" , "IL10" , "MX1",  "MX2", "ISG15", "ISG20", "OAS1", "OAS2", "IL4", "IL5", "IL13", "GATA3")

DotPlot(seurat_object, features = poo) +RotatedAxis()
```