---
title: "Candida_sc_Analysis.Rmd"
author: "Nathan Morris"
date: "2024-01-05"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = getwd())
```

### Data Import
```{r echo = TRUE, include= TRUE, error = FALSE, warning = FALSE, message=FALSE}
# Data Import
library(Seurat)
library(scater)
library(dplyr)
library(SingleR)
library(celldex)
library(scRNAseq)
library(scran)
library(gridExtra)
library(pheatmap)
library(ess)
library(reshape2)
library(ggpubr)
library(scCustomize)
library(gridExtra)
library(monocle3)

data_dir <- "/Users/johnmorris/Desktop/Comp_Bio_MS_Weill_Cornell/Summer_2023/Betel Lab/Betel:Iliev Candida Project/Candida_Migratory_TCR_Project/Data/"

setwd(data_dir)
```

```{r}
seurat_object <- readRDS(file = paste0(data_dir, "dataset_clean_scRNA_ADT_new_metadata.Rds"))
```

```{r}
Idents(object = seurat_object) <- "RNA_clusters"

pa <- DimPlot(seurat_object, reduction = "umap_rna", label = TRUE, repel = TRUE) + ggtitle("RNA UMAP with RNA Clusters") + NoLegend()
pb <- DimPlot(seurat_object, reduction = "umap_adt", label = TRUE, repel = TRUE) + ggtitle("Protein UMAP with RNA Clusters") + NoLegend()

Idents(object = seurat_object) <- "Protein_clusters"

pc <- DimPlot(seurat_object, reduction = "umap_rna", label = TRUE, repel = TRUE) + ggtitle("RNA UMAP with Protein Clusters") + NoLegend()
pd <- DimPlot(seurat_object, reduction = "umap_adt", label = TRUE, repel = TRUE) + ggtitle("Protein UMAP with Protein Clusters") + NoLegend()


grid.arrange(pa, pb, pc, pd, ncol = 2)
```

# 1. RNA ADT Cluster Comparison
```{r}
compare_cluster1 <- seurat_object$Protein_clusters
compare_cluster2 <- seurat_object$RNA_clusters

cluster_intersection <- matrix(nrow = length(unique(compare_cluster1)), ncol = length(unique(compare_cluster2)))
cluster_union <- matrix(nrow = length(unique(compare_cluster1)), ncol = length(unique(compare_cluster2)))

for (i in seq(1:nrow(cluster_intersection))){
  cluster.i <- i - 1
  numi <- compare_cluster1 == cluster.i
  for (j in seq(1:ncol(cluster_intersection))){
    cluster.j <- j - 1
    numj <- compare_cluster2 == cluster.j
    cluster_intersection[i,j] <- sum(numi == TRUE & numj == TRUE)
    cluster_union[i,j] <- sum(numi == TRUE | numj == TRUE)
  }
}

cluster_matrix <- cluster_intersection / cluster_union
rownames(cluster_matrix) <- paste("Protein", sort(unique(compare_cluster1)))
colnames(cluster_matrix) <- paste("RNA", sort(unique(compare_cluster2)))

new_matrix <- cbind(cluster_matrix[,"RNA 3"], cluster_matrix[,"RNA 2"], cluster_matrix[,"RNA 1"], cluster_matrix[,"RNA 0"], cluster_matrix[,"RNA 4"], cluster_matrix[,"RNA 5"], cluster_matrix[,"RNA 6"], cluster_matrix[,"RNA 10"], cluster_matrix[,"RNA 7"], cluster_matrix[,"RNA 11"], cluster_matrix[,"RNA 12"], cluster_matrix[,"RNA 9"], cluster_matrix[,"RNA 13"], cluster_matrix[,"RNA 14"], cluster_matrix[,"RNA 15"],cluster_matrix[,"RNA 16"] )
colnames(new_matrix) <- c("RNA 3","RNA 2","RNA 1","RNA 0","RNA 4","RNA 5","RNA 6","RNA 10","RNA 7","RNA 11", "RNA 12", "RNA 9","RNA 13","RNA 14","RNA 15","RNA 16")

pheatmap(new_matrix, main = "Cluster Similarity Heatmap between RNA and Protein Clusters", legend = TRUE, cluster_rows = FALSE, cluster_cols = FALSE, legend_labels = c("0", "0.2", "0.4", "0.6", "0.8", "Intersect over Union"), legend_breaks = c(0, 0.2, 0.4, 0.6, 0.8, max(new_matrix)) )
```

# 2. ## Reclustering with Higher Resolution

```{r}
seurat_object <- FindNeighbors(object = seurat_object, reduction = "umap_rna", dims = 1:2)
seurat_object <- FindClusters(seurat_object_high_clust, resolution = c(0.2, 0.4, 0.6, 0.8, 1.0, 1.2), cluster.name = "high_clust")
```

```{r}
seurat_object$New_RNA_clust <- seurat_object$RNA_snn_res.0.6
```

```{r}
Idents(object = seurat_object) <- "RNA_snn_res.0.6"
DimPlot(seurat_object, reduction = "umap_rna", label = TRUE, repel = TRUE)
```

```{r}
DotPlot(seurat_object, assay = "RNA", features = c("Il17a", "Rora"), group.by = "RNA_snn_res.0.6", split.by = "Infection") + theme(axis.text.x = element_text(angle = 45, hjust = 1)) + coord_flip()

VlnPlot(seurat_object, assay = "RNA", features = c("Il17a"), group.by = "RNA_snn_res.0.6", split.by = "Infection") + theme(axis.text.x = element_text(angle = 45, hjust = 1))
```

```{r}
VlnPlot(seurat_object, assay = "RNA", features = c("Il17a"), group.by = "RNA_snn_res.0.6")

VlnPlot(seurat_object[,seurat_object$RNA_snn_res.0.6 %in% c("15", "19")], assay = "RNA", features = c("Il17a"), group.by = "RNA_snn_res.0.6", split.by = "Infection")
```

```{r}
table(seurat_object[,seurat_object$RNA_snn_res.0.6 %in% c("19")]$Infection)
table(seurat_object[,seurat_object$RNA_snn_res.0.6 %in% c("19")]$Organ)
```


# NEW 01_05_2024

# 2.A ## Identifying Th17 cells

Obviously shows that some cells that express Il17a are hard to find in the 
```{r}
table(seurat_object$Infection[which(seurat_object@assays[["RNA"]]@counts["Il17a",] > 1)])
```


```{r}
il17_cytokines <- c("Il17a", "Il17f", "Il21", "Il22")
il17_tfs <- c("Rorgt")
il17_receptors <- c("CD161", "CCR4", "CCR6")
  
AverageExpression(seurat_object, assay = "RNA", features = c("Il17a", "Rora"), group.by = "New_RNA_clust")

AverageExpression(seurat_object, assay = "RNA", features = c("Il17a", "Rora"), group.by = c("New_RNA_clust", "Infection"))

Stacked_VlnPlot(seurat_object, features = Th17, group.by = "RNA_new_inf")
```

```{r}
FeaturePlot(seurat_object, features = "Il17a")
```

```{r}
Idents(seurat_object) <- "New_RNA_clust"
DefaultAssay(object = seurat_object) <- "RNA"

seurat_object.RNAmarkers_new <- FindAllMarkers(seurat_object, only.pos = TRUE, min.pct = 0.10, logfc.threshold = 0.25)

seurat_object.RNAmarkers_new_top5 <- Extract_Top_Markers(marker_dataframe = seurat_object.RNAmarkers_new, num_genes = 5, data_frame = TRUE,
    rank_by = "avg_log2FC", named_vector = FALSE, make_unique = TRUE)

seurat_object <- ScaleData(seurat_object, features = rownames(seurat_object))

DoHeatmap(seurat_object, features = c(unique(seurat_object.RNAmarkers_new_top5$gene))) #+ coord_flip()

Clustered_DotPlot(seurat_object, features = c(unique(seurat_object.RNAmarkers_new_top5$gene))) #+ coord_flip()

```

```{r}

Idents(seurat_object) <- "New_RNA_clust"
DefaultAssay(object = seurat_object) <- "ADT"

seurat_object.ADTmarkers <- FindAllMarkers(seurat_object, only.pos = TRUE, min.pct = 0.10, logfc.threshold = 0.25)
```

## Cluster 15 - RORalpha very high

https://pubmed.ncbi.nlm.nih.gov/18164222/

Cluster 15 has high Ror alpha, but very low levels of Il17a. It is known that Rorg helps T cell differentiation into Th-17 cells, however this paper states that Ror alpha also helps the differentiation into Th-17 cells.

Could Cluster 15 be helping the proliferation of Cluster 19 / Il17a?

"We report Th17 cells highly expressed another related nuclear receptor, ROR alpha, induced by transforming growth factor-beta and interleukin-6 (IL-6), which is dependent on signal transducer and activator of transcription 3."

Cluster 15 also has high Il2


Therefore, 

# 3. Cell composition for each cluster by organ

```{r}
cluster_organ_comp <- as.data.frame(matrix(nrow = length(unique(seurat_object$RNA_clusters)), ncol = 4))
colnames(cluster_organ_comp) <- c("Cluster", "Lung %", "Colon %", "mLns %" )
#unique(seurat_object$RNA_clusters)
for (i in 0:16){
  cluster_organ_comp[i+1,"Cluster"] = i
  #table(seurat_object$Organ[which(seurat_object$RNA_clusters == i)])
  cluster_organ_comp[i+1,"Lung %"] <- ((length(which(seurat_object$Organ[which(seurat_object$RNA_clusters == i)] == "Lung")))/(length(seurat_object$Organ[which(seurat_object$RNA_clusters == i)])))
  cluster_organ_comp[i+1,"Colon %"] <- ((length(which(seurat_object$Organ[which(seurat_object$RNA_clusters == i)] == "Colon")))/(length(seurat_object$Organ[which(seurat_object$RNA_clusters == i)])))
  cluster_organ_comp[i+1,"mLns %"] <- ((length(which(seurat_object$Organ[which(seurat_object$RNA_clusters == i)] == "mLns")))/(length(seurat_object$Organ[which(seurat_object$RNA_clusters == i)])))
}

df_long <- reshape2::melt(cluster_organ_comp, id.vars = "Cluster", variable.name = "Organ", value.name = "Percentage")

ggplot(df_long, aes(x = Cluster, y = Percentage, fill = Organ)) +
  geom_bar(stat = "identity", position = "dodge") +
  labs(x = "Cluster", y = "Percentage") +
  scale_fill_manual(values = c("Lung %" = "blue", "Colon %" = "green", "mLns %" = "red")) +
  theme_minimal() + ggtitle("Cluster Organ Composition") + scale_x_continuous(breaks=seq(1,16,1))
```

The above shows us that the clusters are each mostly dominant by one or maybe two organ cells. Therefore, the T cells, even if sorted from each organ individually, are still somehow phenotypically associated with their organ of reference.


# 4. Visualizing Infection Differential Abundance on UMAP

```{r}
Idents(object = seurat_object) <- "Infection"
DimPlot(seurat_object, reduction = "umap_rna", label = FALSE, repel = TRUE) +  ggtitle("RNA UMAP by Infection")

Idents(object = seurat_object) <- "RNA_clusters"

DimPlot(seurat_object, reduction = "umap_rna", label = FALSE, repel = TRUE, split.by = "Infection") +  ggtitle("RNA UMAP split by Infection") + labs(color = "RNA Cluster")
```

The RNA UMAP grouped by infection doesn't show any real distinction in any cluster.

The RNA UMAP split by the Infection. As you can see the cluster in the middle (cluster 5) seems to have differential abundance between the Candida and non-infected population, however no new cluster forms when the infection happens.

## 4.A Visualizing Canonical Marker genes and Proteins

```{r}
# Feature Plots using the WNN UMAP
# Visualizing the expression of canonical marker genes and proteins on the multimodal UMAP

# How to search for a gene name
#seurat_object[["RNA"]][grepl("Tcf", rownames(seurat_object[["RNA"]])),][,1]

# Gene Groups from Dr. Iliev's Papers
gois <- c("Tcf7", "Slamf6", "Ccr7", "Cxcr6", "Foxp3", "Mki67", "Klf2") # from RNA assay
Tbet <- c("Tbx21")
stem_like_gois <- c("Ccr7", "Il7r", "Sell", "Tcf7", "Tcf1", "Slamf6", "Klf3", "Il16ra")
# "Smurf1"
activation_like_gois <- c("Cd44", "Cd69", "Il17a", "Il17f", "Il22", "Ifng", "Mki67", "Csf2")
ADT_markers_of_interest <- list("CD8A", "CD4", "CD3", "CCR4.CD194", "CCR6", "CD19") # from Cite-seq Assay

neuro_related_gois <- c("Calca", "Ramp1", "Ramp3", "Ntrk3", "Gch1", "Nr3c1", "Bex3")

DefaultAssay(object = seurat_object) <- "RNA"
Idents(object = seurat_object) <- "RNA_clusters"
seurat_object <- ScaleData(seurat_object)

# Expression Plots
FeaturePlot(seurat_object, features = c(gois),
                  reduction = 'umap_rna', 
                  cols = c("lightgrey","darkgreen"), ncol = 3)

FeaturePlot(seurat_object, features = c(ADT_markers_of_interest),
                  reduction = 'umap_rna', 
                  cols = c("lightgrey","darkgreen"), ncol = 3)

FeaturePlot(seurat_object, features = c(activation_like_gois),
                  reduction = 'umap_rna', max.cutoff = 2, 
                  cols = c("lightgrey","darkgreen"), ncol = 3)

#RidgePlot(seurat_object, features = c(stem_like_gois), ncol = 2)

DotPlot(seurat_object, features = c(gois)) + RotatedAxis() + ggtitle("RNA Clusters by Th17 stem-like markers")

DoHeatmap(seurat_object, features = c(gois)) + RotatedAxis() + ggtitle("RNA Clusters by Th17 stem-like markers")
```

Here, I'm using the genes that Dr. Iliev's papers point to as genes that direct the differentiation of Th17 cells. 

Cxcr6, Foxp3 and Mki67 seem to cluster relatively well. Cxcr6 is closely related to cluster 5, and Foxp3 in clusters 7 and 9. Mki67 is expressed a lot in cluster 11. The cite-seq for the CD8A and CD4 antibodies seem to show the two different T cell populations distinctly. This is found in other datasources as well (https://www.sciencedirect.com/science/article/pii/S0092867421005833), where CD8 and CD4 are partially blended together when analyzing the transcriptome but separated clearly in protein data.


```{r}
DefaultAssay(object = seurat_object) <- "ADT"
ADT_markers <- c("CD62L", "CCR6", "CD4", "CD8A")

FeaturePlot(seurat_object, features = c(ADT_markers),
                  reduction = 'umap_rna', 
                  cols = c("lightgrey","darkgreen"), ncol = 2)

like_markers <- c("Rora","Il17a","Ccr6", "Il23r")
DefaultAssay(object = seurat_object) <- "RNA"
FeaturePlot(seurat_object, features = c(like_markers),
                  reduction = 'umap_rna', 
                  cols = c("lightgrey","darkgreen"), ncol = 2)
```


# 4.B UMAP Marker Genes

```{r}
DefaultAssay(seurat_object) <- "ADT"

p1 <- FeaturePlot(seurat_object, features = c("CD62L"),
            reduction = 'umap_rna', 
            cols = c("yellow","red"), ncol = 4) + NoLegend() + NoAxes()
p2 <- FeaturePlot(seurat_object, features = c("CD4"),
            reduction = 'umap_rna', 
            cols = c("yellow","red"), ncol = 4) + NoLegend() + NoAxes()
p3 <- FeaturePlot(seurat_object, features = c("CD8A"),
            reduction = 'umap_rna', 
            cols = c("yellow","red"), ncol = 4) + NoLegend() + NoAxes()
p4 <- FeaturePlot(seurat_object, features = c("CCR6"),
            reduction = 'umap_rna', 
            cols = c("yellow","red"), ncol = 4) + NoLegend() + NoAxes()

grid.arrange(p1,p2,p3,p4, ncol = 2)
```

```{r}
VlnPlot(seurat_object, assay = "ADT", features = c("CD62L", "CD4", "CD8A", "CCR6"), group.by = "RNA_clusters", ncol = 2)
```

```{r}
DefaultAssay(seurat_object) <- "RNA"

p1 <- FeaturePlot(seurat_object, features = c("Rorc"),
            reduction = 'umap_rna', 
            cols = c("yellow","red"), ncol = 4) + NoLegend() + NoAxes()
p2 <- FeaturePlot(seurat_object, features = c("Il17a"),
            reduction = 'umap_rna', 
            cols = c("yellow","red"), ncol = 4) + NoLegend() + NoAxes()
p3 <- FeaturePlot(seurat_object, features = c("Ccr6"),
            reduction = 'umap_rna', 
            cols = c("yellow","red"), ncol = 4) + NoLegend() + NoAxes()
p4 <- FeaturePlot(seurat_object, features = c("Il23r"),
            reduction = 'umap_rna', 
            cols = c("yellow","red"), ncol = 4) + NoLegend() + NoAxes()

grid.arrange(p1,p2,p3,p4, ncol =4)
```


### Reclustering by organ

Here, I wanted to see if separating the seurat object by organ would allow us to see individual clusters better. And perhaps be able to identify different clusters of interest from these.

There really wasn't much distinction. Therefore, I decided to not have this code run-able.
```{r}
organs <- c("Colon", "Lung", "mLns")

seurat_object.colon <- subset(x = seurat_object, subset = (Organ == "Colon"))
seurat_object.lung <- subset(x = seurat_object, subset = (Organ == "Lung"))
seurat_object.mlns <- subset(x = seurat_object, subset = (Organ == "mLns"))
```


```
# Colon
seurat_object.colon <- FindMultiModalNeighbors(seurat_object.colon, reduction.list = list("umap_rna", "umap_adt"), dims.list = list(1:2,1:2), modality.weight.name = list("WNN.weight_RNA", "WNN.weight_ADT"))
seurat_object.colon <- RunUMAP(seurat_object.colon, nn.name = "weighted.nn", reduction.name = "wnn.umap", reduction.key = "wnnUMAP_")
seurat_object.colon <- FindClusters(seurat_object.colon, graph.name = "wsnn", algorithm = 3, resolution = 1, verbose = FALSE)
unique(seurat_object.colon@meta.data[["wsnn_res.1"]]) # Using resolution = 1 --> Has 10 total clusters

# Lung
seurat_object.lung <- FindMultiModalNeighbors(seurat_object.lung, reduction.list = list("umap_rna", "umap_adt"), dims.list = list(1:2,1:2), modality.weight.name = list("WNN.weight_RNA", "WNN.weight_ADT"))
seurat_object.lung <- RunUMAP(seurat_object.lung, nn.name = "weighted.nn", reduction.name = "wnn.umap", reduction.key = "wnnUMAP_")
seurat_object.lung <- FindClusters(seurat_object.lung, graph.name = "wsnn", algorithm = 3, resolution = 1, verbose = FALSE)
unique(seurat_object.lung@meta.data[["wsnn_res.1"]]) # Using resolution = 1 --> Has 16 total clusters

# mLns
seurat_object.mlns <- FindMultiModalNeighbors(seurat_object.mlns, reduction.list = list("umap_rna", "umap_adt"), dims.list = list(1:2,1:2), modality.weight.name = list("WNN.weight_RNA", "WNN.weight_ADT"))
seurat_object.mlns <- RunUMAP(seurat_object.mlns, nn.name = "weighted.nn", reduction.name = "wnn.umap", reduction.key = "wnnUMAP_")
seurat_object.mlns <- FindClusters(seurat_object.mlns, graph.name = "wsnn", algorithm = 3, resolution = 1, verbose = FALSE)
unique(seurat_object.mlns@meta.data[["wsnn_res.1"]]) # Using resolution = 1 --> Has 14 total clusters
```

```
Idents(object = seurat_object.colon) <- "WNN_clusters_UMAP"
DimPlot(seurat_object.colon, reduction = "wnn.umap", label = TRUE, repel = TRUE) +  ggtitle("Colon WNN UMAP by Cluster")
#WNN.colon <- FindAllMarkers(seurat_object.colon, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)

#Idents(object = seurat_object.colon) <- "Immgen.fine.RNA"
Idents(object = seurat_object.colon) <- "Infection"
DimPlot(seurat_object.colon, reduction = "wnn.umap", label = TRUE, repel = TRUE) +  ggtitle("Colon WNN UMAP by Infection")

Idents(object = seurat_object.lung) <- "WNN_clusters_UMAP"
DimPlot(seurat_object.lung, reduction = "wnn.umap", label = TRUE, repel = TRUE) +  ggtitle("Lung WNN UMAP by Cluster")
#WNN.lung <- FindAllMarkers(seurat_object.lung, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)

#Idents(object = seurat_object.lung) <- "Immgen.fine.RNA"
Idents(object = seurat_object.lung) <- "Infection"
DimPlot(seurat_object.lung, reduction = "wnn.umap", label = TRUE, repel = TRUE) +  ggtitle("Lung WNN UMAP by Infection")

Idents(object = seurat_object.mlns) <- "WNN_clusters_UMAP"
DimPlot(seurat_object.mlns, reduction = "wnn.umap", label = TRUE, repel = TRUE) +  ggtitle("mLns WNN UMAP by Cluster")
#WNN.mlns <- FindAllMarkers(seurat_object.mlns, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)

#Idents(object = seurat_object.mlns) <- "Immgen.fine.RNA"
Idents(object = seurat_object.mlns) <- "Infection"
DimPlot(seurat_object.mlns, reduction = "wnn.umap", label = TRUE, repel = TRUE) +  ggtitle("mLns WNN UMAP by Infection")
```

Interesting clusters based on genes of interest - 
Colon - cluster 3 (Rora, Ccr6, Cxcr6, ) - Could be a T-stem-like, cluster 5 - foxp3 - Could be a T-reg set
Lung - cluster 4 (Foxp3, rora, Ccr6) - Looks CD8
mLns - cluster 6 (Foxp3) - Could be a T-reg set

However, based on the Infection UMAPs from each organ, I don't see any clusters of interest, or clusters arising due to the infection.

# 5. Differential Abundance Proptest - Candida infected cell # vs Non-infected # for all organs and WNN_UMAP clusters


```{r, warning = FALSE}
cluster_differences <- function(cluster_name){
  
  organs <- c("Colon", "Lung", "mLns")
  
  df_organ <- c()
  df_cluster <- c()
  df_NT_inClust <- c()
  df_CA_inClust <- c()
  df_NT_total <- c()
  df_CA_total <- c()
  df_pcNT <- c()
  df_pcCA <- c()
  df_pval <- c()
  
  n <- 1
  
  RNA_organ <- c(seurat_object.colon, seurat_object.lung, seurat_object.mlns)
  cluster_name <- as.character(cluster_name)
  
  for (seurat_object.sub in RNA_organ) {
    
    organ <- as.character(unique(seurat_object.sub$Organ))
    clusters = sort(unique(seurat_object.sub[[cluster_name]][[1]]))
    
    for (clust in clusters) {
  
      NTc <- nrow(seurat_object.sub@meta.data[seurat_object.sub[[cluster_name]][[1]] == clust & seurat_object.sub@meta.data$Infection == "No", ])
      # NTc is the number of no Candida from an organ in a particular individual cluster
      CAc <- nrow(seurat_object.sub@meta.data[seurat_object.sub[[cluster_name]][[1]] == clust & seurat_object.sub@meta.data$Infection == "Ca", ])
      # CAc is the number of Candida from an organ in a particular individual cluster
      
      NT <- nrow(seurat_object.sub@meta.data[seurat_object.sub@meta.data$Infection == "No", ])
      # Total amount of No infection from the organ
      CA <- nrow(seurat_object.sub@meta.data[seurat_object.sub@meta.data$Infection == "Ca", ])
      # Total amount of Ca infection from the organ
      
      successes <- c(NTc, CAc)
      trials <- c(NT, CA)
      
      result <- tryCatch({
    res <- prop.test(x = successes, n = trials)
      }, warning = function(w) {
    # Code to handle warnings
    warningCondition <- conditionMessage(w)
    cat("Warning:", warningCondition, "\n")
      })
      # Prop test to test the differences between the number of No/Ca in that cluster
      # Versus the total number of those samples from the organ
      
      df_organ[n] <- organ
      df_cluster[n] <- clust
      df_NT_inClust[n] <- NTc
      df_CA_inClust[n] <- CAc
      df_NT_total[n] <- NT
      df_CA_total[n] <- CA
      df_pcNT[n] <- NTc/NT
      df_pcCA[n] <- CAc/CA
      df_pval[n] <- res$p.value
      
      n <- n + 1
    }
  }  
  df <- data.frame(Organ = df_organ, Cluster = df_cluster, NT_clust = df_NT_inClust, CA_clust = df_CA_inClust, NT_total = df_NT_total, CA_total = df_CA_total, NT_Proportion = df_pcNT, CA_Proportion = df_pcCA, p_value = df_pval)
  df$adj_p_value <- p.adjust(df$p_value, method = "bonferroni")
  return(df)
}
```

```{r}
Idents(seurat_object.colon) <- 'RNA_clusters'
Idents(seurat_object.mlns) <- 'RNA_clusters'
Idents(seurat_object.lung) <- 'RNA_clusters'
#df <- cluster_differences('RNA_clusters')


df_new <- cluster_differences('New_RNA_clust')


#write.csv(df, "RNA_by_Organ_Candida_diff.csv")
```

By subsetting by organ first, we can see the significant levels of 

```{r}
df[df$adj_p_value < 0.05,]
```

```{r}
#Colon Cluster 3
#rev(sort(table(seurat_object.colon$Immgen.fine.RNA[seurat_object.colon$WNN_clusters_UMAP == 3])))
rev(sort(table(seurat_object.colon$RNA_clusters[seurat_object.colon$RNA_clusters == 3])))

# Lung Cluster 4
#rev(sort(table(seurat_object.lung$Immgen.fine.RNA[seurat_object.lung$WNN_clusters_UMAP == 4])))
rev(sort(table(seurat_object.lung$RNA_clusters[seurat_object.lung$RNA_clusters == 5])))
```

From our RNA clusters, only cluster 5 is the cluster with significant difference in Ca and Non-infection composition in multiple organs (colon and lung). 

Looking here at which RNA-cluster these cells are apart of, you can see they are mostly in RNA cluster 5.

Cluster 0 and 2 are significantly different in the colon, and are close to being significant in the lung. This is primarily due to the composition switch in Clusters 0 and 2 between the cells.


# 5.A-1 Cluster CA/NT Proportion Visual

```{r}
#library(reshape2)
#df.long <- melt(df, id.vars=c("Organ", "Cluster"))
library(tidyr)
df.long <- df_new %>%
  pivot_longer(cols = c(NT_Proportion, CA_Proportion),
               names_to = "Proportion_Type",
               values_to = "Proportion_Value")

ggplot(df.long, aes(x = Proportion_Type, y = Proportion_Value, fill = Organ)) +
  geom_bar(stat = "identity", position = "dodge") +
  facet_wrap(~Cluster, scales = "free_y") +
  labs(x = "Cluster", y = "Proportion")

my_comparisons <- list(c("NT_Proportion", "CA_Proportion"))

ggplot(df.long, aes(x = Organ, y = Proportion_Value, fill = Proportion_Type)) +
  geom_bar(stat = "identity", position = "dodge") +
  facet_wrap(~Cluster, scales = "free_y") +
  labs(x = "Cluster", y = "Proportion") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  theme(legend.position = "top") + stat_compare_means(comparisons = my_comparisons, method = "wilcox.test", label =  "p.signif", paired = FALSE) #+ ylim(0,0.5)
```

Here, NT Proportion and CA proportion, are showing for each organ, what percentage of all Infected cells are in that cluster, and what percentage of all non-Infected cells are in that cluster.

Therefore the number for cluster 0 for colon =  the number of colon cells infected in cluster 0 / the number of colon cells infected in total The number of colon cells infected in cluster 0 is the number of successes, while the number of colon cells infected in total is the number of trials.

Then the prop.test is the the number of colon cells infected in cluster 0 / the number of colon cells infected in total and the number of colon cells non-infected in cluster 0 / the number of colon cells not infected in total

Significant diff # of infected and non-infected in each organ by RNA_cluster shown below
```
df[df$adj_p_value < 0.05,]
```

```{r}
temp <- as.data.frame(melt(df))
temp <- temp[temp$variable %in% c("NT_Proportion", "CA_Proportion"),]
```

```{r}
my_comparisons <- list(c("NT_Proportion", "CA_Proportion"))

ggplot(temp, aes(x = Organ, y = value, color = variable)) +
  geom_boxplot() +
  labs(
    x = "Cluster",
    y = "Frequency",
    title = "Boxplot of Frequency by Cluster and Condition"
  ) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) + facet_grid(. ~ Cluster) +
  theme(legend.position = "top") + stat_compare_means(comparisons = my_comparisons, method = "t.test", label =  "p.signif", paired = FALSE) + ylim(0,0.5)
```

```{r}
temp <- na.omit(temp)

# Create the plot
ggplot(temp, aes(x = Organ, y = value, color = variable)) +
  geom_boxplot() +
  labs(
    x = "Cluster",
    y = "Frequency",
    title = "Boxplot of Frequency by Cluster and Condition"
  ) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  facet_grid(. ~ Cluster) +
  theme(legend.position = "top") +
  stat_compare_means(
    comparisons = list(c("CA_Proportion", "NT_Proportion")),
    method = "wilcox.test", # You can choose a different statistical test
    method.args = list(alternative = "two.sided"),
    label = "p.format"
  )
```

RNA_cluster 5 - Dr. Iliev was primarily concerned with cluster 5, based on the statistically different cell counts in Candida mice vs non-Infected mice for colon and lung samples.

```{R}
df <- as.data.frame(table(seurat_object$Mouse_New, seurat_object$RNA_clusters))
colnames(df) <- c("Mouse", "Cluster", "Frequency")

my_comparisons <- list(c("No Infection Male", "Infected Male"),
                       c("Infected Male", "Infected Female"),
                       c("No Infection Male", "Infected Female"))

library(ggh4x)

plot <- ggplot(df, aes(x = Mouse, y = Frequency, color = Mouse)) +
  geom_boxplot() +
  labs(
    x = "Cluster",
    y = "Frequency",
    title = "Boxplot of Frequency by Cluster and Condition"
  ) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) + facet_grid(. ~ Cluster, scales = "free_y") +
  theme(legend.position = "top") + stat_compare_means(comparisons = my_comparisons, method = "t.test", label =  "p.signif", paired = FALSE)

#plot + facetted_pos_scales(
#    y = list(
#      Cluster == c(5:12) ~ scale_y_continuous(breaks = c(0, 500)) 
#  ))
```

## 5.B For Takato -  Cluster Organ Composition Visual for Takato

```{r}
df_organ_clust <- as.data.frame(table(seurat_object$Organ, seurat_object$Infection, seurat_object$RNA_clusters))

colnames(df_organ_clust) <- c("Organ", "Infection", "Cluster", "Freq")
```

```{r}
library(ggplot2)

# Assuming 'df' is your dataframe
# Convert Cluster and Organ to factors to control the order on the plot
df_organ_clust$Cluster <- factor(df_organ_clust$Cluster)
df_organ_clust$Organ <- factor(df_organ_clust$Organ)

df_organ_clust <- df_organ_clust %>%
  group_by(Cluster, Infection) %>%
  mutate(Clust_Proportion = Freq / sum(Freq))

df_organ_clust <- df_organ_clust %>%
  group_by(Organ, Infection) %>%
  mutate(Organ_Proportion = Freq / sum(Freq))

df_organ_clust$Organ_Infection <- paste0(df_organ_clust$Organ, " ", df_organ_clust$Infection)

df_organ_clust$Clust_Infection <- paste0(df_organ_clust$Cluster, " ", df_organ_clust$Infection)

df_organ_clust$Clust_Infection <- factor(df_organ_clust$Clust_Infection, levels = rev(c("0 Ca", "0 No", "1 Ca", "1 No", "2 Ca", "2 No", "3 Ca", "3 No", "4 Ca", "4 No", "5 Ca", "5 No", "6 Ca", "6 No", "7 Ca", "7 No", "8 Ca", "8 No", "9 Ca", "9 No", "10 Ca", "10 No", "11 Ca", "11 No", "12 Ca", "12 No", "13 Ca", "13 No", "14 Ca", "14 No", "15 Ca", "15 No", "16 Ca", "16 No")))
```

### Figure for Organ Make Up in Each Cluster/Infection
```{r}
# Using Facet
ggplot(df_organ_clust, aes(x = Infection, y = Clust_Proportion, fill = Organ)) +
  geom_bar(stat = "identity") +
  facet_wrap(~Cluster, scales = "free_y", ncol = 1, , switch = 'y', ) +  # Separate facets for each infection status
  labs(title = "Proportion of Each Organ in Each Cluster and Infection Status",
       x = "Infection",
       y = "Proportion",
       fill = "Organ") + coord_flip() + #theme_minimal() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) # Rotate x-axis labels for better visibility

# No Facet
ggplot(df_organ_clust, aes(x = Clust_Infection, y = Clust_Proportion, fill = Organ)) +
  geom_bar(stat = "identity") +
  labs(title = "Proportion of Each Organ in Each Cluster and Infection Status",
       x = "Cluster",
       y = "Proportion",
       fill = "Organ") + coord_flip() + #  theme_minimal() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))  # Rotate x-axis labels for better
```


## Figure for Cluster Make Up in Each Organ/Infection
```{r}
# Using Facet
ggplot(df_organ_clust, aes(x = Infection, y = Organ_Proportion, fill = Cluster)) +
  geom_bar(stat = "identity") +
  facet_wrap(~Organ, scales = "free_y", ncol = 1, , switch = 'y', ) +  # Separate facets for each infection status
  labs(title = "Proportion of Each Cluster in Each Organ and Infection Status",
       x = "Infection and Organ",
       y = "Proportion",
       fill = "Cluster") + coord_flip() + #theme_minimal() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) # Rotate x-axis labels for better visibility

# No Facet
ggplot(df_organ_clust, aes(x = Organ_Infection, y = Organ_Proportion, fill = Cluster)) +
  geom_bar(stat = "identity") +
  labs(title = "Proportion of Each Cluster in Each Organ and Infection Status",
       x = "Infection and Organ",
       y = "Proportion",
       fill = "Cluster") + coord_flip() + #  theme_minimal() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))  # Rotate x-axis labels for better visibility 
```


Note: This is how much of each clusters with or without infection is made up of organs. Therefore, each cluster/infection combination has 100% when adding each organ together.

However, the differential abundance analysis instead uses what proportion of cells from each Organ and Non/Infected cells are in each Cluster. Therefore, each organ/infection combination has 100% when adding all clusters together.

TEST: As a test, I want to see if we use the first way, do we get similar results

```
res_df <- matrix(nrow = length(unique(df_organ_clust$Cluster)), ncol = length(unique(df_organ_clust$Organ)))
rownames(res_df) <- unique(df_organ_clust$Cluster)
colnames(res_df) <- unique(df_organ_clust$Organ)

for (cluster in unique(df_organ_clust$Cluster)){
  for (organ in unique(df_organ_clust$Organ)){
    print(cluster)
    print(organ)
    
    sub_df_clust <- df_organ_clust[df_organ_clust$Cluster == cluster,]
    sub_df_clust_organ <- df_organ_clust[df_organ_clust$Cluster == cluster & df_organ_clust$Organ == organ,]

    #ca_prop <- c(sub_df_clust_organ$Freq[sub_df_clust_organ$Infection == "Ca"] ,sum(sub_df_clust$Freq[sub_df_clust$Infection == "Ca"]))
    #no_prop <- c(sub_df_clust_organ$Freq[sub_df_clust_organ$Infection == "No"] ,sum(sub_df_clust$Freq[sub_df_clust$Infection == "No"]))
    
    clust_prop <- c(sub_df_clust_organ$Freq[sub_df_clust_organ$Infection == "Ca"] ,sub_df_clust_organ$Freq[sub_df_clust_organ$Infection == "No"])
    total_prop <- c(sum(sub_df_clust$Freq[sub_df_clust$Infection == "Ca"]),sum(sub_df_clust$Freq[sub_df_clust$Infection == "No"]))
    
    res <- prop.test(x = clust_prop, n = total_prop)
    
    res_df[cluster, organ] <- res$p.value
  }
}

# Adjusted p-value with 48 tests would be 0.05/48  =  0.001041667
```

## Other Visualization by Mouse and Organ

Before with the proportions, I was checking the amount of cells within each cluster, with or without infection, divided by the 

```{r}
seurat_object.colon <- subset(x = seurat_object, subset = (Organ == "Colon"))
seurat_object.lung <- subset(x = seurat_object, subset = (Organ == "Lung"))
seurat_object.mlns <- subset(x = seurat_object, subset = (Organ == "mLns"))
```

```{r}
#Colon
colon_mouse_cell_freq <- as.data.frame(table(seurat_object.colon$Mouse_New, seurat_object.colon$RNA_clust_annot))
colnames(colon_mouse_cell_freq) <- c("Mouse", "Cell_Type", "Frequency")
# Lung
lung_mouse_cell_freq <- as.data.frame(table(seurat_object.lung$Mouse_New, seurat_object.lung$RNA_clust_annot))
colnames(lung_mouse_cell_freq) <- c("Mouse", "Cell_Type", "Frequency")
#mLns
mlns_mouse_cell_freq <- as.data.frame(table(seurat_object.mlns$Mouse_New, seurat_object.mlns$RNA_clust_annot))
colnames(mlns_mouse_cell_freq) <- c("Mouse", "Cell_Type", "Frequency")

mice <- unique(seurat_object$Mouse_New)

for (mouse in mice){
  colon_mouse_cell_freq$Freq_Perc[colon_mouse_cell_freq$Mouse == mouse] <- colon_mouse_cell_freq$Frequency[colon_mouse_cell_freq$Mouse == mouse] / sum(seurat_object.colon$Mouse_New == mouse)
  
  lung_mouse_cell_freq$Freq_Perc[lung_mouse_cell_freq$Mouse == mouse] <- lung_mouse_cell_freq$Frequency[lung_mouse_cell_freq$Mouse == mouse] / sum(seurat_object.lung$Mouse_New == mouse)
  
  mlns_mouse_cell_freq$Freq_Perc[mlns_mouse_cell_freq$Mouse == mouse] <- mlns_mouse_cell_freq$Frequency[mlns_mouse_cell_freq$Mouse == mouse] / sum(seurat_object.mlns$Mouse_New == mouse)
}
```

```{r}
lung_mouse_cell_freq$Mouse <- factor(lung_mouse_cell_freq$Mouse)
colon_mouse_cell_freq$Mouse <- factor(colon_mouse_cell_freq$Mouse)
mlns_mouse_cell_freq$Mouse <- factor(mlns_mouse_cell_freq$Mouse)

# Create the stacked barplot using ggplot2 and facet by 'Mouse'
p1 <- ggplot(lung_mouse_cell_freq, aes(x = Cell_Type, y = Freq_Perc, fill = Cell_Type)) +
  geom_bar(stat = "identity", position = "stack") +
  labs(x = NULL, y = "Proportion") + #coord_flip() +
  facet_wrap(~Mouse) + 
  theme(axis.text.y = element_text(angle = 0, hjust = 0.5), legend.position="none", axis.text.x=element_blank()) + ggtitle("Lung Cell Type Frequency by Mouse")

#legend.position="top"

p2 <- ggplot(colon_mouse_cell_freq, aes(x = Cell_Type, y = Freq_Perc, fill = Cell_Type)) +
  geom_bar(stat = "identity", position = "stack") +
  labs(x = NULL, y = "Proportion") + #coord_flip() +
  facet_wrap(~Mouse) + 
  theme(axis.text.y = element_text(angle = 0, hjust = 0.5), legend.position="none", axis.text.x=element_blank()) + ggtitle("Colon Cell Type Frequency by Mouse")

p3 <- ggplot(mlns_mouse_cell_freq, aes(x = Cell_Type, y = Freq_Perc, fill = Cell_Type)) +
  geom_bar(stat = "identity", position = "stack") +
  labs(x = NULL, y = "Proportion") + #coord_flip() +
  facet_wrap(~Mouse) + 
  theme(axis.text.x = element_text(angle = 65, hjust = 0.5), legend.position="none") + ggtitle("mlns Cell Type Frequency by Mouse")


#ggplot(lung_mouse_cell_freq, aes(x = Mouse, y = Freq_Perc, fill = Cell_Type)) +
#  geom_bar(stat = "identity", position = "stack") +
#  labs(x = NULL, y = "Proportion") +
#  #facet_wrap(~Mouse) + 
#  theme(axis.text.x = element_text(angle = 45, hjust = 1)) + ggtitle("Lung Cell Type Frequency by Mouse")
```

```{r, fig.height = 10}
#p1 + p2 + p3
plist <- list(p1,p2,p3)
grid.arrange(grobs = plist, nrow = 3)
```

Some interesting findings from this:

There is barely any NK CCl5 or Treg in the mouse0080, non-infected male, in the Lung, while there is a very high proportion in both of the others.

Gzma Memory T (Cluster 13) is not found in the mouse0082 at all.






# 6. Top 5 Genes for Cluster
```
# Find markers for all RNA clusters

Idents(seurat_object) <- "RNA_clusters"
DefaultAssay(object = seurat_object) <- "RNA"

seurat_object.RNAmarkers <- FindAllMarkers(seurat_object, only.pos = TRUE, min.pct = 0.10, logfc.threshold = 0.25)

saveRDS(seurat_object.RNAmarkers, file = paste0(data_dir,"/migratory_T_cell_cluster_dge.rds"), ascii = FALSE, version = NULL,
        compress = TRUE, refhook = NULL)

# Extract Top Genes
result <- seurat_object.RNAmarkers %>%
  arrange(cluster, p_val_adj) %>%
  group_by(cluster) %>% slice_head(n=5)

migratory_t_cell_top_5_df <- Extract_Top_Markers(marker_dataframe = seurat_object.RNAmarkers, num_genes = 5, data_frame = TRUE,
    rank_by = "avg_log2FC", named_vector = FALSE, make_unique = TRUE)

saveRDS(migratory_t_cell_top_5_df, file = paste0(data_dir,"/migratory_T_cell_cluster_top5_genes.rds"), ascii = FALSE, version = NULL,
        compress = TRUE, refhook = NULL)
```

```{r}
migratory_t_cell_marker_v2 <- readRDS(paste0(data_dir,"/migratory_T_cell_cluster_dge.rds"))
migratory_t_cell_top_5_df <- readRDS(paste0(data_dir,"/migratory_T_cell_cluster_top5_genes.rds"))
```

```{r}
DefaultAssay(seurat_object) <- "RNA"
Idents(object = seurat_object) <- "RNA_clusters"
Clustered_DotPlot(seurat_object, features = c(unique(migratory_t_cell_top_5_df$gene), "Cd4", "Sell", "Il17a")) #+ coord_flip()

DefaultAssay(seurat_object) <- "RNA"
seurat_object <- ScaleData(seurat_object, features = rownames(seurat_object))
DoHeatmap(seurat_object, features = c(unique(migratory_t_cell_top_5_df$gene), "Cd4", "Sell")) #+ coord_flip()

DoHeatmap(seurat_object, features = c(unique(migratoryt_cell_marker_v2$gene[migratory_t_cell_marker_v2$cluster == 5]), "Cd4", "Sell")) #+ coord_flip()

```

### Checking Genes for Each Cluster

Cluster 0 - Weird how it is enriched in the Non-infected mouse, rather than infected

Junb - AP-1 transcription factor, shown to be essential for the development of Il-17 producing cells
https://www.nature.com/articles/s41385-019-0182-0.pdf

GPR183 - Migration and Activation of Immune cells on B cells, T cells and DCs
https://bpspubs.onlinelibrary.wiley.com/doi/epdf/10.1111/bph.15311

CCR7 - Lymphocyte development and dendritic cell homing to th lymph nodes
A ligand receptor to help the homing of dendritic cells and T cells, in order to establish antigen-specific activation
https://www.nature.com/articles/nri2297

Ly6c1 - Supports preferential homing of central memory CD8+ T cells into lymph nodes
https://pubmed.ncbi.nlm.nih.gov/21308682/

Tcf7 - Transcription factor 7 - encodes for TCF-1 protein
https://translational-medicine.biomedcentral.com/articles/10.1186/s12967-015-0617-7

FeatureScatter(pbmc3k.final, feature1 = "LYZ", feature2 = "CCL5")
Cluster 1:
Fam241a
Ccr7
Cd8b1
Cd8a
Klf21

# 6.A Top 5 Proteins for Cluster

```
# Find markers for all RNA clusters

Idents(seurat_object) <- "RNA_clusters"
DefaultAssay(object = seurat_object) <- "ADT"

seurat_object.ADTmarkers <- FindAllMarkers(seurat_object, only.pos = TRUE, min.pct = 0.10, logfc.threshold = 0.25)

saveRDS(seurat_object.ADTmarkers, file = paste0(data_dir,"/migratory_T_cell_cluster_dpe.rds"), ascii = FALSE, version = NULL,
        compress = TRUE, refhook = NULL)

```

```
# Extract Top Genes
result <- seurat_object.ADTmarkers %>%
  arrange(cluster, p_val_adj) %>%
  group_by(cluster) %>% slice_head(n=5)

migratory_t_cell_top_5_df <- Extract_Top_Markers(marker_dataframe = seurat_object.RNAmarkers, num_genes = 5, data_frame = TRUE,
    rank_by = "avg_log2FC", named_vector = FALSE, make_unique = TRUE)

saveRDS(migratory_t_cell_top_5_df, file = paste0(data_dir,"/migratory_T_cell_cluster_top5_genes.rds"), ascii = FALSE, version = NULL,
        compress = TRUE, refhook = NULL)
```

```{r}
migratory_t_cell_protein_marker_v2 <- readRDS(paste0(data_dir,"/migratory_T_cell_cluster_dpe.rds"))
#migratory_t_cell_top_5_df <- readRDS(paste0(data_dir,"/migratory_T_cell_cluster_top5_genes.rds"))
```

```{r}
DefaultAssay(seurat_object) <- "ADT"
Idents(object = seurat_object) <- "RNA_clusters"
Clustered_DotPlot(seurat_object, features = c(unique(migratory_t_cell_protein_marker_v2$gene))) #+ coord_flip()

DefaultAssay(seurat_object) <- "ADT"
seurat_object <- ScaleData(seurat_object, features = rownames(seurat_object))
DoHeatmap(seurat_object, features = c(unique(migratory_t_cell_protein_marker_v2$gene)))

```


# 7. Clusters by Description Genes / Markers

```{r}
DefaultAssay(seurat_object) <- "RNA"

Global_T <- c("Cd3e","Cd3d")
CD4_8 <- c("Cd4", "Cd8a")
Th17 <- c("Rora","Il26","Il17a","Il17f","Ccr6", "Il23r", "Il22")
Tfh_Th1_Th2_Th17 <- c("Cd200","Ptpn13","Btla","Tbx21","Gata3","Cd40lg", "Il2")
Tregs <- c("Foxp3","Il2ra","Ctla4","Ikzf2")
Naive <- c("Il7r","Ccr7","Tcf7","Sell","Lef1")
Memory <- c("Eomes","Gzmk","Cxcr3")
Eff_Mem <- c("Fgfbp2","Klrd1")
Activated <- c("Tnf","Ifng","Fos","Jun")
#Exhausted_activated <- c("Lag3","Havcr2","Pdcd1")
Exhausted <- c("Gzmb","Entpd1","Itgae")
NK <- c("Tyrobp","Ncam1","Cd160", "Nkg7")
Neuron <- c("Calca", "Ramp1", "Ramp3", "Ntrk3", "Gch1", "Nr3c1", "Bex3")

#Th17 <- c("Il17a","Il17f","Ccr6", "Rorc")
#Tfh <- c("Cxcr5", "Bcl6", "Cxcr3", "Pdcd1","Cd40lg")
#NK <- c("Tyrobp","Ncam1","Cd160","Gnly", "Nkg7","Klrb1a", "Gzma", "Cxcr6")

All_genes <- list(Global_T,CD4_8,Th17,Tfh_Th1_Th2_Th17,Tregs,Naive,Memory,Eff_Mem,Activated,Exhausted, NK, Neuron)
names(All_genes) <- c("Global_T","CD4/8","Th17","Tfh_Th1_Th2","Tregs","Naive","Memory","Eff_Mem","Activated","Exhausted", "NK", "Neuron")

DotPlot(
    seurat_object, 
    features = All_genes, 
    scale = TRUE, 
    group.by = "RNA_clusters") + RotatedAxis() + NoLegend()


DotPlot(
    seurat_object, 
    features = All_genes, 
    scale = TRUE, 
    group.by = "New_RNA_clust") + RotatedAxis() + NoLegend()

DotPlot(
    seurat_object, 
    features = All_genes, 
    scale = TRUE, 
    group.by = "New_RNA_clust", split.by = "Infection") + RotatedAxis() + NoLegend()



DotPlot(
    seurat_object[,seurat_object$New_RNA_clust %in% c("15", "19", "22", "27")], 
    features = All_genes, 
    scale = TRUE, 
    group.by = "New_RNA_clust", split.by = "Infection") + RotatedAxis() + NoLegend()


DotPlot(
    seurat_object, 
    features = All_genes, 
    scale = TRUE, 
    group.by = "RNA_clust_annot",
    cols = c("blue", "white", "red")
  ) + RotatedAxis() #+ NoLegend()

Clustered_DotPlot(
    seurat_object, 
    features = unlist(All_genes),
    group.by = "RNA_clusters"
  ) + RotatedAxis() + NoLegend()

#Clustered_DotPlot(seurat_object, features = c(unique(migratory_t_cell_top_5_df$gene)

DefaultAssay(seurat_object) <- "RNA"
Idents(object = seurat_object) <- "RNA_clusters"
marker_dotplot <- lapply(names(All_genes), function(x) {
  DotPlot(
    seurat_object, 
    features = All_genes[[x]], 
    scale = TRUE, 
    col.min = 0, 
    col.max = 5, 
    group.by = "RNA_clusters"
  ) + RotatedAxis() + labs(title = x) + NoLegend()
})

grid.arrange(grobs = marker_dotplot, nrow = 2)
```

## Showing the TCRGD vs the TCRB

```{r}
DefaultAssay(seurat_object) <- "ADT"

DotPlot(seurat_object, features = c("TCRB", "TCRGD"), scale = TRUE, group.by = "New_RNA_clust") + RotatedAxis() + NoLegend()
```

Comments- 
https://www.nature.com/articles/s41598-018-35161-5
AverageExpression(seurat_object, features = c("Tbx21", "Ifng", ), group.by = "RNA_clusters")
Tbx21 is very high in the cluster 16 ()
Ifng is high in cluster 13 (Th1)
Rorc is high in cluster 12
Il10 is very high in cluster 9


https://perspectivesinmedicine.cshlp.org/content/5/3/a019612

## 7.A Annotate Clusters based on Markers

```
Idents(object = seurat_object) <- "RNA_clusters"

seurat_object <- RenameIdents(seurat_object,
                              `0` = 'Naive CD4 - Lung',
                              `1` = 'Naive CD8 - Lung',
                              `2` = 'Naive CD8 - mLns',
                              `3` = 'Naive CD4 - mLns', 
                              `4` = 'CD8 Eff Mem - Colon',
                              `5` = 'Th-17',
                              `6` = 'NK Ccl5',
                              `7` = 'Treg Tnfrsf4',
                              `8` = 'Cluster 8',
                              `9` = 'Treg Tnfrsf4',
                              `10` = 'CD8 Ly6c2',
                              `11` = 'Cluster 11',
                              `12` = 'NK Tmem/Il17r',
                              `13` = 'Gzma Memory T - Lung',
                              `14` = 'CD8 Isg15 - mLns',
                              `15` = 'Naive Lef1 - mLns',
                              `16` = 'Gzma Exhausted - Lung')

seurat_object[["RNA_clust_annot"]] <- as.factor(Idents(seurat_object))

DimPlot(seurat_object, reduction = "umap_rna", group.by = "RNA_clust_annot", label = TRUE)
```

```{r}
Idents(object = seurat_object) <- "RNA_clusters"

seurat_object <- RenameIdents(seurat_object,
                              `0` = 'Naive CD4',
                              `1` = 'Naive CD8',
                              `2` = 'Naive CD8',
                              `3` = 'Naive CD4', 
                              `4` = 'CD8 Eff Mem - Colon',
                              `5` = 'Th-17',
                              `6` = 'NK Ccl5',
                              `7` = 'Treg Tnfrsf4',
                              `8` = 'Cluster 8',
                              `9` = 'Treg Tnfrsf4',
                              `10` = 'CD8 Ly6c2',
                              `11` = 'Cluster 11',
                              `12` = 'TCR-gd Il17r',
                              `13` = 'Gzma Memory T - Lung',
                              `14` = 'CD8 Isg15 - mLns',
                              `15` = 'Naive Lef1 - mLns',
                              `16` = 'Gzma Exhausted - Lung')

seurat_object[["RNA_clust_annot"]] <- as.factor(Idents(seurat_object))

DimPlot(seurat_object, reduction = "umap_rna", group.by = "RNA_clust_annot", label = TRUE)
```

# 8. Interleukins for Takato

```{r}
il_genes <- sort(grep("^Il", rownames(seurat_object@assays[["RNA"]]), value = TRUE))

il_genes_short <- c("Il6","Il23a", "Il23r","Il17a", "Il10", "Il16", "Il12a", "Il13", "Il21", "Il22", "Il2", "Il3", "Il4", "Il5", "Il7", "Il9")

seurat_object$RNA_inf <- paste0(seurat_object$RNA_clusters,"-", seurat_object$Infection)

DotPlot(seurat_object, assay = "RNA", features = il_genes[0:50], group.by = "RNA_inf") + theme(axis.text.x = element_text(angle = 45, hjust = 1))

DotPlot(seurat_object, assay = "RNA", features = il_genes[50:90], group.by = "RNA_inf") + theme(axis.text.x = element_text(angle = 45, hjust = 1))

DotPlot(seurat_object, assay = "RNA", features = il_genes_short, group.by = "RNA_inf") + theme(axis.text.x = element_text(angle = 45, hjust = 1))
```
  
```{r}
seurat_object$RNA_new_inf <- paste0(seurat_object$New_RNA_clust,"-", seurat_object$Infection)

DotPlot(seurat_object, assay = "RNA", features = il_genes[0:50], group.by = "RNA_new_inf") + theme(axis.text.x = element_text(angle = 45, hjust = 1))

DotPlot(seurat_object, assay = "RNA", features = il_genes[50:90], group.by = "RNA_new_inf") + theme(axis.text.x = element_text(angle = 45, hjust = 1))

DotPlot(seurat_object, assay = "RNA", features = il_genes_short, group.by = "RNA_new_inf") + theme(axis.text.x = element_text(angle = 45, hjust = 1))
```
  
  
```
VlnPlot(seurat_object, features = c(neuro_related_gois), ncol = 3)
AverageExpression(seurat_object, features = c(neuro_related_gois), ncol = 3)

FeaturePlot(seurat_object, features = c(neuro_related_gois), reduction = 'umap_rna', cols = c("yellow","red"), ncol = 3)
```

# 8. Neuro Related Genes


```{r}
neuro_related_gois <- c("Calca", "Ramp1", "Ramp3", "Ntrk3", "Gch1", "Nr3c1", "Bex3")


VlnPlot(seurat_object, features = c(neuro_related_gois), ncol = 3)
AverageExpression(seurat_object, features = c(neuro_related_gois), ncol = 3)

FeaturePlot(seurat_object, features = c(neuro_related_gois), reduction = 'umap_rna', cols = c("yellow","red"), ncol = 3)
```

```{r}
VlnPlot(seurat_object[,seurat_object$RNA_clusters == "5"], features = c(neuro_related_gois), group.by = "Infection")

AverageExpression(seurat_object[,seurat_object$RNA_clusters == "5"], features = c(neuro_related_gois), group.by = "Infection")

VlnPlot(seurat_object[,seurat_object$RNA_clusters == "12"], features = c(neuro_related_gois), group.by = "Infection")

AverageExpression(seurat_object[,seurat_object$RNA_clusters == "12"], features = c(neuro_related_gois), group.by = "Infection")
```

# 8. Figures for Takato

# 8.A UMAP Splits

UMAP Split by Infection and 
```{r}
Idents(object = seurat_object) <- "Infection"

p_all <- DimPlot(seurat_object, reduction = "umap_rna", label = TRUE, repel = TRUE, group.by = "RNA_clust_annot") + ggtitle("RNA UMAP with Cluster Annotations") + NoLegend()
p_inf <- DimPlot(seurat_object, reduction = "umap_rna", label = TRUE, repel = TRUE, split.by = "Infection", group.by = "RNA_clust_annot") + ggtitle("RNA UMAP by Infection") + NoLegend()
p_organ <- DimPlot(seurat_object, reduction = "umap_rna", label = TRUE, repel = TRUE, split.by = "Organ", group.by = "RNA_clust_annot") + ggtitle("RNA UMAP by Organ") + NoLegend()

p_all + p_inf / p_organ

```

````{r}
p_all <- DimPlot(seurat_object, reduction = "umap_rna", label = TRUE, repel = TRUE, group.by = "RNA_clust_annot") + ggtitle("RNA UMAP with Cluster Annotations") + NoLegend()
p_inf <- DimPlot(seurat_object, reduction = "umap_rna", label = TRUE, repel = TRUE, split.by = "Infection", group.by = "RNA_clust_annot") + ggtitle("RNA UMAP by Infection") + NoLegend()
p_organ <- DimPlot(seurat_object, reduction = "umap_rna", label = TRUE, repel = TRUE, split.by = "Organ", group.by = "RNA_clust_annot") + ggtitle("RNA UMAP by Organ") + NoLegend()

p_all / p_inf + p_organ
```

# 8.B 



# 9. CytoTrace
```{r}
library(CytoTRACE)
data_tmp <- as.matrix(GetAssayData(object = seurat_object , slot = "counts", assay="RNA"))
results <- CytoTRACE(data_tmp, subsamplesize = 1000)
phenotype = as.character(seurat_object$RNA_clust_annot)
names(phenotype) <- rownames(seurat_object[[]])
plotCytoTRACE(results, phenotype = phenotype, emb=seurat_object[["umap_rna"]]@cell.embeddings)
```


## CytoTRACE
```{r}
library(CytoTRACE)
data_tmp <- as.matrix(GetAssayData(object = seurat_object, slot = "counts", assay="RNA"))
results <- CytoTRACE(data_tmp, subsamplesize = 1000)
phenotype = as.character(seurat_object$RNA_clust_annot)
names(phenotype) <- rownames(seurat_object[[]])
plotCytoTRACE(results, phenotype = phenotype, emb=seurat_object[["umap_rna"]]@cell.embeddings)

plotCytoTRACE(results, phenotype = phenotype, emb=seurat_object[["umap_rna"]]@cell.embeddings, gene = "Sell")
```

# 10. Pseudotime Analysis

I've decided it's best to do it for CD4 and CD8 cells for each individual organ.
```{r}
#devtools::install_github("satijalab/seurat-wrappers")
library(monocle3)
```

```{r}
DefaultAssay(seurat_object) <- "RNA"

seurat_object.colon <- subset(x = seurat_object, subset = (Organ == "Colon"))
seurat_object.lung <- subset(x = seurat_object, subset = (Organ == "Lung"))
seurat_object.mlns <- subset(x = seurat_object, subset = (Organ == "mLns"))
```

Here, I'm dividing the clusters with a higher resolution. Here, we are hoping to get the cluster 
```{r}
#cds <- SeuratWrappers::as.cell_data_set(seurat_object.mlns)
cds <- SeuratWrappers::as.cell_data_set(seurat_object[,seurat_object$Mouse == "mouse0080"])

cds
# colData(cds) # Metadata here
#fData(cds) # Gene Metadata (empty)
#rownames(fData(cds)) # Gene names
# counts(cds) # Gives counts
fData(cds)$gene_short_name <- rownames(fData(cds)) # Requires short names
#fData(cds)$gene_short_name <- rownames(cds) # Requires short names


recreate.partition <- c(rep(1,length(cds@colData@rownames)))
names(recreate.partition) <- cds@colData@rownames
recreate.partition <- as.factor(recreate.partition)

cds@clusters$UMAP$partitions <- recreate.partition

#RNA_clusters <- seurat_object.mlns$RNA_clust_annot
RNA_clusters <- seurat_object.mlns$RNA_clusters

cds@clusters$UMAP$clusters <- RNA_clusters

# Assign UMAP coordinate - cell embeddings

#cds@int_colData@listData$reducedDims$UMAP<- seurat_object.mlns@reductions$umap_rna@cell.embeddings

cds@int_colData@listData$reducedDims$UMAP<- seurat_object[,seurat_object$Mouse == "mouse0080"]@reductions$umap_rna@cell.embeddings

#pData(cds)$sample_name <- NULL

#cluster.before.trajectory <- plot_cells(cds, color_cells_by = "cluster", label_groups_by_cluster = FALSE, group_label_size = 5) + theme(legend.position = "right")

pData(cds)$sample_name <- as.factor(pData(cds)$sample_name)

pData(cds) <- NULL

cds <- learn_graph(cds, use_partition = FALSE)

plot_cells(cds,
           color_cells_by = "cluster",
           label_groups_by_cluster = FALSE,
           label_branch_points = FALSE,
           label_roots = TRUE,
           label_leaves = FALSE,
           group_label_size = 5)
```

Ordering by pseudotime - not sure what to make the root cells
```{r}
#cds <- order_cells(cds, reduction_method = "UMAP", root_cells = colnames(cds[,clusters(cds) %in% c("Naive CD4", "Naive CD8")]))
#cds@clusters@listData[["UMAP"]][["clusters"]] %in% c("Naive CD4", "Naive CD8")
cds <- order_cells(cds, reduction_method = "UMAP", root_cells = cds@colData@rownames[cds@clusters@listData[["UMAP"]][["clusters"]] %in% c("Naive CD4", "Naive CD8")])

cds <- order_cells(cds, reduction_method = "UMAP", root_cells = cds@colData@rownames[cds@clusters@listData[["UMAP"]][["clusters"]] %in% c(0,1,2,3)])

# I would want to use wnn.UMAP, but gives an error

plot_cells(cds,
           color_cells_by = "pseudotime",
           label_groups_by_cluster = FALSE,
           label_branch_points = FALSE,
           label_roots = FALSE,
           label_leaves = FALSE,
           group_label_size = 5)

cds$monocle3_pseudotime <- pseudotime(cds)
data.pseudo <- as.data.frame(colData(cds))
#ggplot(data.pseudo, aes(monocle3_pseudotime, RNA_clusters, fill = RNA_clusters)) + geom_boxplot()
#ggplot(data.pseudo, aes(monocle3_pseudotime, Immgen.main.RNA, fill = Immgen.main.RNA)) + geom_boxplot()
ggplot(data.pseudo, aes(monocle3_pseudotime, RNA_clust_annot, fill = RNA_clust_annot)) + geom_boxplot()
```

# 10. CellChatDB


# May want to use CellPhoneDB first
https://github.com/ventolab/CellphoneDB/blob/master/notebooks/DEGs_calculation/0_prepare_your_data_from_Seurat.ipynb

```
library(CellChat)
library(patchwork)
options(stringsAsFactors = FALSE)

https://htmlpreview.github.io/?https://github.com/sqjin/CellChat/blob/master/tutorial/CellChat-vignette.html
```

```
DefaultAssay(seurat_object) <- "RNA"
meta <- seurat_object@meta.data
cellchat <- createCellChat(object = seurat_object, group.by = "RNA_clust_annot")


CellChatDB <- CellChatDB.mouse # use CellChatDB.mouse if running on mouse data
showDatabaseCategory(CellChatDB)

CellChatDB.use <- CellChatDB

# How to fix Issue identified!! Please check the official Gene Symbol of the following genes: H2-BI H2-Ea-ps 
which(CellChatDB.use[["interaction"]]$ligand == "H2-BI") # 1887
CellChatDB.use[["interaction"]] <- CellChatDB.use[["interaction"]][-1887,]
which(CellChatDB.use[["interaction"]]$ligand == "H2-Ea-ps") #1900
CellChatDB.use[["interaction"]] <- CellChatDB.use[["interaction"]][-1900,]


cellchat@DB <- CellChatDB.use

cellchat <- subsetData(cellchat)

cellchat <- identifyOverExpressedGenes(cellchat)
cellchat <- identifyOverExpressedInteractions(cellchat)

cellchat <- computeCommunProb(cellchat)

cellchat <- computeCommunProbPathway(cellchat)

cellchat <- aggregateNet(cellchat)
```

```
saveRDS(cellchat, file = paste0(data_dir,"/cellchat_db.rds"), ascii = FALSE, version = NULL,
        compress = TRUE, refhook = NULL)
```

```{r}
cellchat <- readRDS(paste0(data_dir,"/cellchat_db.rds"))
```

```{r}
groupSize <- as.numeric(table(cellchat@idents))
par(mfrow = c(1,2), xpd=TRUE)
netVisual_circle(cellchat@net$count, vertex.weight = groupSize, weight.scale = T, label.edge= F, title.name = "Number of interactions")
netVisual_circle(cellchat@net$weight, vertex.weight = groupSize, weight.scale = T, label.edge= F, title.name = "Interaction weights/strength")


mat <- cellchat@net$weight
par(mfrow = c(4,4), xpd=TRUE)
for (i in 1:nrow(mat)) {
  mat2 <- matrix(0, nrow = nrow(mat), ncol = ncol(mat), dimnames = dimnames(mat))
  mat2[i, ] <- mat[i, ]
  netVisual_circle(mat2, vertex.weight = groupSize, weight.scale = T, edge.weight.max = max(mat), title.name = rownames(mat)[i])
}
```


## TGFB Signaling

```{r}
pathways.show <- c("TGFb")
# Hierarchy plot
# Here we define `vertex.receive` so that the left portion of the hierarchy plot shows signaling to fibroblast and the right portion shows signaling to immune cells 
vertex.receiver = seq(1,4) # a numeric vector. 
netVisual_aggregate(cellchat, signaling = pathways.show,  vertex.receiver = vertex.receiver)
# Circle plot
par(mfrow=c(1,1))
netVisual_aggregate(cellchat, signaling = pathways.show, layout = "circle")
#netVisual_heatmap(cellchat, signaling = pathways.show, color.heatmap = "Reds")

netAnalysis_contribution(cellchat, signaling = pathways.show)
```

```{r}
netVisual_bubble(cellchat, sources.use = 4, targets.use = c(0:14), remove.isolate = FALSE)
netVisual_bubble(cellchat, sources.use = 4, targets.use = c(0:14), signaling = c("TGFb"), remove.isolate = FALSE)
```

When looking at only the TFGB pathway for the CellChatDB output, we can see that there is a lot of interaction between Cluster 11 and itself, the Th-17 cluster and Cluster 11 and the Treg cluster.

It is possible there is a connection between Th-17 cluster and Cluster 11, therefore Th-17 differentiation is communicating to the proliferating cluster to proliferate in the lung.

```
cellchat <- netAnalysis_computeCentrality(cellchat, slot.name = "netP") 
#netAnalysis_signalingRole_network(cellchat, signaling = pathways.show, width = 8, height = 2.5, font.size = 10)
```
Options for pathways
````{r}
sort(unique(cellchat@DB[["interaction"]][["pathway_name"]]))
```

# IFN-g or Interferon-gamma signaling

```{r}
pathways.show <- c("IFN-II") # TGFb # IFN-II # IL-16
# Hierarchy plot
# Here we define `vertex.receive` so that the left portion of the hierarchy plot shows signaling to fibroblast and the right portion shows signaling to immune cells 
vertex.receiver = seq(1,4) # a numeric vector. 
netVisual_aggregate(cellchat, signaling = pathways.show,  vertex.receiver = vertex.receiver)
# Circle plot
par(mfrow=c(1,1))
netVisual_aggregate(cellchat, signaling = pathways.show, layout = "circle")
#netVisual_heatmap(cellchat, signaling = pathways.show, color.heatmap = "Reds")

netAnalysis_contribution(cellchat, signaling = pathways.show)
```

If we look at the Interferon-gamma type 2 mediated signaling, we can see primarily the interactions are between the Gamma-delta T cells (Cluster 12) with Cluster 13, and then Cluster 13 with the Naive Clusters.

Is it possible that the interferon gamma signaling is reducing the proliferation of the gamma-delta T cells because they are not responsive to the fungal infection?

https://bmcinfectdis.biomedcentral.com/articles/10.1186/1471-2334-14-166

## IL-16 Signaling

```{r}
pathways.show <- c("IL6") # TGFb # IFN-II # IL-16
# Hierarchy plot
# Here we define `vertex.receive` so that the left portion of the hierarchy plot shows signaling to fibroblast and the right portion shows signaling to immune cells 
vertex.receiver = seq(1,4) # a numeric vector. 
netVisual_aggregate(cellchat, signaling = pathways.show,  vertex.receiver = vertex.receiver)
# Circle plot
par(mfrow=c(1,1))
netVisual_aggregate(cellchat, signaling = pathways.show, layout = "circle")
#netVisual_heatmap(cellchat, signaling = pathways.show, color.heatmap = "Reds")

netAnalysis_contribution(cellchat, signaling = pathways.show)
```

## MHC-II Signaling

```{r}
pathways.show <- c("MHC-II") # TGFb # IFN-II # IL-16
# Hierarchy plot
# Here we define `vertex.receive` so that the left portion of the hierarchy plot shows signaling to fibroblast and the right portion shows signaling to immune cells 
vertex.receiver = seq(1,4) # a numeric vector. 
netVisual_aggregate(cellchat, signaling = pathways.show,  vertex.receiver = vertex.receiver)
# Circle plot
par(mfrow=c(1,1))
netVisual_aggregate(cellchat, signaling = pathways.show, layout = "circle")
#netVisual_heatmap(cellchat, signaling = pathways.show, color.heatmap = "Reds")

netAnalysis_contribution(cellchat, signaling = pathways.show)
```

## MHC-I Signaling

```{r}
pathways.show <- c("MHC-I") # TGFb # IFN-II # IL-16
# Hierarchy plot
# Here we define `vertex.receive` so that the left portion of the hierarchy plot shows signaling to fibroblast and the right portion shows signaling to immune cells 
vertex.receiver = seq(1,4) # a numeric vector. 
netVisual_aggregate(cellchat, signaling = pathways.show,  vertex.receiver = vertex.receiver)
# Circle plot
par(mfrow=c(1,1))
netVisual_aggregate(cellchat, signaling = pathways.show, layout = "circle")
#netVisual_heatmap(cellchat, signaling = pathways.show, color.heatmap = "Reds")

netAnalysis_contribution(cellchat, signaling = pathways.show)
```

# 10. Investigating Cluster 11 - Proliferating Cells in the Lung

```{r}
Idents(seurat_object) <- "RNA_clusters"
ca_markers_11_ADT <- FindMarkers(seurat_object, ident.1 = "11", only.pos = TRUE,  min.pct = 0.05,logfc.threshold = 0.05, assay = "ADT")
ca_markers_11_RNA <- FindMarkers(seurat_object, ident.1 = "11", only.pos = TRUE,  min.pct = 0.05,logfc.threshold = 0.05, assay = "RNA")
```

From the ADT Markers - CD11A and CD49d are cell surface integrins CD11a and CD49d can be used to identify total antigen-specific CD4 T cell responses to a specific peptide during murine infection with LCMV
https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5551405/

CD11a in CD8+ T-cell accumulation and activation in adipose tissue
https://pubmed.ncbi.nlm.nih.gov/24158516/

or possibly activated NK cells - https://pubmed.ncbi.nlm.nih.gov/32335843/

RNA - 
Stmn1 - macrophages - should be a decrease in Stmn1 when activated https://www.jbc.org/article/S0021-9258(20)42294-X/fulltext

Stathmin - Extensive in vitro studies have strongly suggested that stathmin could act by sequestering tubulin and/or by binding to microtubule tips
This could mean that Stathmin upregulation has an impact on the tubulin make up of the immune cells
https://faseb.onlinelibrary.wiley.com/doi/abs/10.1096/fj.201500125R

Microtubules are important for many cell processes - like cell division, intracellular transport, cell motility, and cell shape

https://www.sciencedirect.com/science/article/pii/S1087184512000953 - Immune cell infiltration by candida alibans could have an impact on some sort of lysis and then impact on the mitotic cell division cycle. Possibly stops the 

RNA markers - include genes associated with cell cycle regulation (Mki67, Birc5, Top2a), chromatin structure (H2ac8, Hmgb2, H1f5, H2az1, Hmgn2), and immune response (Gzma, Lgals1).

Cenpa - CENP-A, the centromeric histone H3 variant, is an epigenetic mark that directly perturbs genetic stability and chromatin when overexpressed
https://www.nature.com/articles/s42003-021-01941-5

H1f5 - Histones H1 are necessary for the condensation of nucleosome chains into higher-order structured fibers. Acts also as a regulator of individual gene transcription through chromatin remodeling, nucleosome spacing and DNA methylation
https://www.uniprot.org/uniprotkb/P16401/entry

```{r}
so_11 <- seurat_object[,seurat_object$RNA_clusters == "12"]
Idents(so_11) <- "Infection"
FeatureScatter(so_11, feature1 = "CD4", feature2 = "CD8A")
```

```{r}
library(AnnotationDbi)
library(org.Mm.eg.db)
library(clusterProfiler)

ca_markers_11_RNA$entrezid <- mapIds(org.Mm.eg.db, keys = rownames(ca_markers_11_RNA), column = "ENTREZID", keytype = "SYMBOL") # Could also use ENSEMBL instead of ENTREZ
ca_markers_11_go_terms_bp <- enrichGO(ca_markers_11_RNA$entrezid, "org.Mm.eg.db", keyType = "ENTREZID", ont = "BP", pvalueCutoff = 0.05, pAdjustMethod = "BH")
dotplot(ca_markers_11_go_terms_bp, showCategory=10) + ggtitle("Cluster 11")

```

GSEA
https://learn.gencore.bio.nyu.edu/rna-seq-analysis/gene-set-enrichment-analysis/

```{r}
original_gene_list_11 <- ca_markers_11_RNA$avg_log2FC
names(original_gene_list_11) <- ca_markers_11_RNA$entrezid
gene_list_11 <- na.omit(original_gene_list_11)
gene_list_11 = sort(gene_list_11, decreasing = TRUE)

gse_11 <- gseGO(geneList=gene_list_11, 
             ont ="ALL", 
             keyType = "ENTREZID", 
             pvalueCutoff = 0.05, 
             verbose = TRUE, 
             OrgDb = org.Mm.eg.db, 
             pAdjustMethod = "none")

require(DOSE)
dotplot(gse_11, showCategory=10, split=".sign") + facet_grid(.~.sign)
```

```{r}
gseaplot(gse_11, by = "all", title = gse$Description[10], geneSetID = 1)
```

```{r}
VlnPlot(seurat_object[,seurat_object$RNA_clusters == "11"], features = "Mki67", group.by = "Infection", split.by = "Organ")
VlnPlot(seurat_object[,seurat_object$RNA_clusters == "11"], features = "Stmn1", group.by = "Infection", split.by = "Organ")
VlnPlot(seurat_object[,seurat_object$RNA_clusters == "11"], features = "Cenpa", group.by = "Infection", split.by = "Organ")
VlnPlot(seurat_object[,seurat_object$RNA_clusters == "11"], features = "H1f5", group.by = "Infection", split.by = "Organ")
```

```{r}
so_11 <- seurat_object[,seurat_object$RNA_clusters == "11"]
Idents(so_12) <- "Infection"
FeatureScatter(so_12, feature1 = "CD62L", feature2 = "CD8A")
```


# 11. Investigating Cluster 12 - gd-T cells in Lung

https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4310324/
γδ T lymphocytes are a minor subset (1–10%) of human peripheral blood T cells. Most (>70%) are CD4−CD8− [double negative (DN)], some (30%) are CD8+CD4− and very few (<1%) are CD4+CD8− [CD8+ or CD4+ single positive (SP), respectively].

```{r}
so_12 <- seurat_object[,seurat_object$RNA_clusters == "12"]
Idents(so_12) <- "Infection"
FeatureScatter(so_12, feature1 = "CD4", feature2 = "CD8A")
```

As we can see here, most of the cells have low expression of both CD4 and CD8. Therefore, I believe they are most likely double negative and therefore gamma-delta T-cells

You can see this also by the lack of the TCRs in the expanded Cluster 12 (highest in the UMAP)

# 11. Investigating Cluster 4 - CD8 Effector in Colon

```{r}
so_4 <- seurat_object[,seurat_object$RNA_clusters == "4"]
Idents(so_4) <- "Infection"
FeatureScatter(so_4, feature1 = "CD62L", feature2 = "CD8A")
```

```{r}
VlnPlot(so_4, features = "Cd7", group.by = "Mouse_New")
```

```{r}
table(so_4$Organ)
```

Cluster 4 is interesting, because it's mainly only contracted in the Colon
And it is significantly higher abundance in the Infected mice vs the Non-infected Mouse

You can also see that the CD62L is low in the Cluster 4, and the CD8A is high in Cluster 4

Therefore, I believe that Cluster 4 is some form of CD8 Memory T cell

https://pubmed.ncbi.nlm.nih.gov/12594257/
CD7 is a differentiation marker that identifies multiple CD8 T cell effector subsets

https://pubmed.ncbi.nlm.nih.gov/33329591/
CD38 can act as an enzyme with NAD-depleting nd intracellular signaling activity

https://www.sciencedirect.com/science/article/pii/S2405654522000774
NAD+ is an enzyme cofactor that is necessary for maintaining cell metabolism

https://www.nature.com/articles/s41467-022-29979-x
Zfp36 influences CD8+ T cell differentiation
ZFP36 and ZFP36L1 function during an early temporal window after activation in CD8+ T cells to limit the tempo of effector differentiation and the functional capacity of differentiated effector cells.


# 11. Investigating Naive Clusters

```{r}
VlnPlot(seurat_object[,seurat_object$RNA_clusters == "1"], features = "VISTA", group.by = "Infection", split.by = "Organ")
```

# 11. Investigating Cluster 13

https://journals.aai.org/jimmunol/article/174/9/5332/72489/Frequency-Specificity-and-Sites-of-Expansion-of
CD8+ T cells in the lungs - increased GzmB frequency reflects a total pool of virus specific cells which have migrated to the lungs, especially early in the response


# Save the Rds
```{r}
#setwd(data_dir)
saveRDS(seurat_object, file = "dataset_post_sc_Analysis.rds")
```