---
title: "Candida_sc_CellChat"
author: "Nathan Morris"
date: "2024-01-09"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = getwd())
```


### Data Import
```{r echo = TRUE, include= TRUE, error = FALSE, warning = FALSE, message=FALSE}
# Data Import
library(Seurat)
library(scater)
library(dplyr)
library(SingleR)
library(celldex)
library(scRNAseq)
library(scran)
library(gridExtra)
library(pheatmap)
library(ess)
library(reshape2)
library(ggpubr)
library(scCustomize)
library(gridExtra)
library(monocle3)
library(reticulate)
```

```{r}
## For Reticulate (Python-R wrapper)
conda_list()
use_condaenv("r-reticulate")    #activate an environment e.g py2_env
py_install("opentsne")
reticulate::py_install("phate", pip=TRUE)
devtools::install_github("KrishnaswamyLab/phateR")
py_install("fa2")
```

# 1. CellChat

```{r}
#devtools::install_github("immunogenomics/presto")
#devtools::install_github("jinworks/CellChat")
library(CellChat)
library(patchwork)
library(NMF)
options(stringsAsFactors = FALSE)

#https://htmlpreview.github.io/?https://github.com/sqjin/CellChat/blob/master/tutorial/CellChat-vignette.html
```

First create separate cell chat objects for infected and non-infected cells
```
seurat_object <- readRDS(file = paste0(data_dir, "/dataset_w_annot.rds"))

DefaultAssay(seurat_object) <- "RNA"
meta <- seurat_object@meta.data

cells.use = rownames(meta)[meta$Infection == "No"]
cells.use = rownames(meta)[meta$Infection == "Ca"]

data.input <- seurat_object[,seurat_object$cell_barcode %in% cells.use]
meta = meta[cells.use, ]

# Create Cell Chat Object
cellchat <- createCellChat(object = data.input, group.by = "CD_annot")

# Add Secreted Signlaing database
CellChatDB.use <- subsetDB(CellChatDB.mouse, search = "Secreted Signaling")
cellchat@DB <- CellChatDB.use

# Subset and pre-processing the expression data
cellchat <- subsetData(cellchat)
cellchat <- identifyOverExpressedGenes(cellchat)
cellchat <- identifyOverExpressedInteractions(cellchat)

# Project Gene Expression data onto protein-protein interaction
cellchat <- projectData(cellchat, PPI.mouse)

# Compute the communication probability and infer cellular communication network
cellchat <- computeCommunProb(cellchat)
cellchat <- filterCommunication(cellchat, min.cells = 10)
cellchat <- computeCommunProbPathway(cellchat)
cellchat <- aggregateNet(cellchat)

# Compute network centrality scores
cellchat <- netAnalysis_computeCentrality(cellchat, slot.name = "netP")
cellchat@net$count
cellchat@net$weight

# Identify and visualize outgoing communication pattern of secreting cells
selectK(cellchat, pattern = "outgoing") 
#nPatterns = 3 # For No
#nPatterns = 3 # For Ca
cellchat <- identifyCommunicationPatterns(cellchat, pattern = "outgoing",
                                      k = nPatterns, width = 5, height = 9)

# Identify and visualize incoming communication pattern of target cells
selectK(cellchat, pattern = "incoming")
#nPatterns = 2 # For No
#nPatterns = 3 # For Ca
cellchat <- identifyCommunicationPatterns(cellchat,pattern = "incoming", 
                                      k = nPatterns, width = 5, height = 9)

# Identify signaling groups based on functional similarity
cellchat <- computeNetSimilarity(cellchat, type = "functional")
cellchat <- netEmbedding(cellchat, type = "functional")
cellchat <- netClustering (cellchat, type = "functional", do.parallel = FALSE)

# Identify signaling groups based on structure similarity
cellchat <- computeNetSimilarity(cellchat, type = "structural")
cellchat <- netEmbedding(cellchat, type = "structural")
cellchat <- netClustering(cellchat, type = "structural", do.parallel = FALSE)
```

```{r}
#saveRDS(cellchat, file = paste0(data_dir, "/cellchat_data/cellchat.No.rds"))
#saveRDS(cellchat, file = paste0(data_dir, "/cellchat_data/cellchat.Ca.rds"))
```

Load both No and Ca object
```{r}
data_dir <- "/Users/johnmorris/Desktop/Comp_Bio_MS_Weill_Cornell/Summer_2023/Betel Lab/Betel:Iliev Candida Project/Candida_Migratory_TCR_Project/Data/"

setwd(data_dir)

cellchat.No <- readRDS(file = paste0(data_dir, "/cellchat_data/cellchat.No.rds"))
cellchat.Ca <- readRDS(file = paste0(data_dir, "/cellchat_data/cellchat.Ca.rds"))

object.list <- list(No = cellchat.No, Ca = cellchat.Ca)

names(object.list)

cellchat <- mergeCellChat(object.list, add.names = names(object.list))

cellchat
```
Lifting Object - Trying
Lifting the object doesn't change the analysis
```
group.new = levels(cellchat.Ca@idents)
cellchat.No_lift <- liftCellChat(cellchat.No, group.new)
object.list_lift <- list(No = cellchat.No_lift, Ca = cellchat.Ca)

names(object.list)

cellchat_lift <- mergeCellChat(object.list_lift, add.names = names(object.list_lift))

cellchat_lift
```


Compare the overall information flow of each signaling pathway
```{r}
comp1 <- rankNet(cellchat, mode = "comparison", stacked = T, measure = "count", do.stat = TRUE) + theme(legend.position="none") 
comp2 <- rankNet(cellchat, mode = "comparison", stacked = T, do.stat = TRUE, measure = "weight")# 42 & 47
comp12 <- comp1 + comp2
pathway_rank <- ggarrange(
  comp1, comp2, labels = c("A", "B"),
  common.legend = TRUE, legend = "bottom"
  )

rankNet(cellchat, mode = "comparison", stacked = F, do.stat = TRUE)
```

# Figure 12

Compare the total number of interactions and interaction strength
```{r}
count_comp <- compareInteractions(cellchat, show.legend = F, group = c(1,2), measure = "count")
weight_comp <- compareInteractions(cellchat, show.legend = F, group = c(1,2), measure = "weight")

total <- ggarrange(
  count_comp, weight_comp, labels = c("C", "D"),
  common.legend = TRUE, legend = "none"
  )

fig_12 <- ggarrange(
  pathway_rank, total,
  common.legend = TRUE, legend = "bottom", ncol = 1)
```

```{r}
fig_directory <- "/Users/johnmorris/Desktop/Comp_Bio_MS_Weill_Cornell/Summer_2023/Betel Lab/Betel:Iliev Candida Project/Candida_Migratory_TCR_Project/Candida_Migratory_TCR/Figures_Tables/"

ggsave(filename = paste0(fig_directory, "figure_12.pdf"), plot = fig_12, device = "pdf", height = 16, width = 16)

ggsave(filename = paste0(fig_directory, "figure_12.jpeg"), plot = fig_12, device = "jpeg", height = 16, width = 16)
```


Compare outgoing/incoming interaction strength for all the cell types
```{r}
count.sum <- sapply(object.list, function(x) {rowSums(x@net$count) + 
    colSums(x@net$count)-diag(x@net$count)})
weight.MinMax <- c(min(count.sum), max(count.sum)) # control the dot size 
gg <- list()
for (i in 1:length(object.list)) {
  gg[[i]] <- netAnalysis_signalingRole_scatter(object.list[[i]], 
  title = names(object.list)[i], weight.MinMax = weight.MinMax)
}

patchwork::wrap_plots(plots = gg)
```

For a specific cell type, you can see the associated signalling changes after infection
```{r}
netAnalysis_signalingChanges_scatter(cellchat, idents.use = "KLRG1+ Eff CD8- Lung")
p1 <- netAnalysis_signalingChanges_scatter(cellchat, idents.use = "KLRG1+ CD8 Eff - Lung",signaling.exclude = "GALECTIN")
p2 <- netAnalysis_signalingChanges_scatter(cellchat, idents.use = "Th17 CD4")
#netAnalysis_signalingChanges_scatter(cellchat, idents.use = "Mix CD4- mLns")
#netAnalysis_signalingChanges_scatter(cellchat, idents.use = "Inflam. DC",signaling.exclude = "MIF")
```

Circle Plot
-Show the number of interactions between any two cell populations 
-Compute the maximum number of cells and the maximum number of interactions among all cell groups
Max number of cells involved is 1214
Max number of interactions among cell groups is 8
```{r}
weight.max <- getMaxWeight(object.list, attribute = c("idents","count"))

par(mfrow = c(1,2), xpd=TRUE)
for (i in 1:length(object.list)) {
  netVisual_circle(object.list[[i]]@net$count, weight.scale = T, label.edge= F,
  edge.weight.max = weight.max[2], edge.width.max = 12, arrow.size = 0.1,
  title.name = paste0("Number of interactions - ", names(object.list)[i]))
}
```


# Circle Plots Supplemental Figure 12

If you wanted to look at a certain signalling pathway use the below
Here I've decided to use the signalling pathways that are upregulated in the Ca condition, IL16 and CCl5
# CCL
```{r}
pathway.show <- c("CCL") # A
weight.max <- getMaxWeight(object.list, slot.name = c("netP"),  attribute =pathway.show)

recorded_plots <- list()

par(mfrow = c(1,2), xpd=TRUE)
for (i in 1:length(object.list)) {
  
  recordPlot()
  
  netVisual_aggregate(object.list[[i]], signaling = pathway.show, layout = "circle",
  edge.weight.max = weight.max[1], edge.width.max = 10, arrow.size = 0.05, 
  signaling.name = paste(pathway.show, names(object.list)[i]))
  
  recorded_plots[[i]] <- recordPlot()
}

ccl_circle <- recorded_plots
```
# IL16
```{r}
pathway.show <- c("IL16") # A
weight.max <- getMaxWeight(object.list, slot.name = c("netP"),  attribute =pathway.show)

recorded_plots <- list()

par(mfrow = c(1,2), xpd=TRUE)
for (i in 1:length(object.list)) {
  
  recordPlot()
  
  netVisual_aggregate(object.list[[i]], signaling = pathway.show, layout = "circle",
  edge.weight.max = weight.max[1], edge.width.max = 10, arrow.size = 0.05, 
  signaling.name = paste(pathway.show, names(object.list)[i]))
  
  recorded_plots[[i]] <- recordPlot()
}

il16_circle <- recorded_plots
```

# IL2

```{r}
il2_circle <- netVisual_aggregate(object.list[[2]], signaling = pathway.show, layout = "circle",
  edge.weight.max = weight.max[1], edge.width.max = 10, arrow.size = 0.05, 
  signaling.name = paste("IL2 Ca signaling pathway network"))
```

# IFN-II

```{r}
ifn_circle <- netVisual_aggregate(object.list[[1]], signaling = pathway.show, layout = "circle",
  edge.weight.max = weight.max[1], edge.width.max = 10, arrow.size = 0.05, 
  signaling.name = paste("IFN-II No signaling pathway network"))
```

# Supplemental Figure 12

```{r}
grid.grabExpr(il16_circle[[1]])

cowplot::plot_grid( cowplot::plot_grid(il16_circle[[2]], labels = "A"), cowplot::plot_grid(ccl_circle[[1]], labels = "B"), nrow = 2)

supp_12a <- cowplot::plot_grid( il16_circle[[2]], labels = c("A"))

supp_12b <- cowplot::plot_grid(ccl_circle[[2]], labels = c("B"))

supp_12c <- cowplot::plot_grid(il2_circle, nrow = 1, labels = c("C"))

supp_12d <- cowplot::plot_grid(ifn_circle, nrow = 1, labels = c("D"))

```


```{r}
ggsave(filename = paste0(fig_directory, "supplemental_12a.pdf"), plot = supp_12a, device = "pdf", height = 20, width = 24, units = "in")

#ggsave(filename = paste0(fig_directory, "supplemental_12a.jpeg"), plot = supp_12a, device = "jpeg", height = 20, width = 24)

ggsave(filename = paste0(fig_directory, "supplemental_12b.pdf"), plot = supp_12b, device = "pdf", height = 20, width = 24, units = "in")

#ggsave(filename = paste0(fig_directory, "supplemental_12b.jpeg"), plot = supp_12b, device = "jpeg", height = 36, width = 30)

ggsave(filename = paste0(fig_directory, "supplemental_12c.pdf"), plot = supp_12c, device = "pdf", height = 20, width = 24)

#ggsave(filename = paste0(fig_directory, "supplemental_12cd.jpeg"), plot = supp_12b, device = "jpeg", height = 36, width = 30)

ggsave(filename = paste0(fig_directory, "supplemental_12d.pdf"), plot = supp_12d, device = "pdf", height = 20, width = 24)
```


# Heatmaps for All Pathways

```{r}
all_pathways <- union(object.list[[1]]@netP$pathways, 
                       object.list[[2]]@netP$pathways)
# Shows we have 11 pathways

ht1 = netAnalysis_signalingRole_heatmap(object.list[[1]], pattern = "all", 
      signaling = all_pathways, title = names(object.list)[1],  
      width = 7, height = 11, color.heatmap = "OrRd")
ht2 = netAnalysis_signalingRole_heatmap(object.list[[2]], pattern = "all", 
      signaling = all_pathways, title = names(object.list)[2], 
      width = 7, height = 11, color.heatmap = "OrRd")
overall <- ComplexHeatmap::draw(ht1 + ht2, ht_gap = unit(0.5, "cm"))
```


Looking at only outgoing patterns
```{r}
ht3 = netAnalysis_signalingRole_heatmap(object.list[[1]], pattern ="outgoing", 
      signaling = all_pathways, title = names(object.list)[1], 
      width = 9, height = 11)
ht4 = netAnalysis_signalingRole_heatmap(object.list[[2]], pattern ="outgoing", 
      signaling = all_pathways, title = names(object.list)[2], 
      width = 9, height = 11)
out <- ComplexHeatmap::draw(ht3 + ht4, ht_gap = unit(0.5, "cm"))
```

Incoming patterns

```{r}
ht5 = netAnalysis_signalingRole_heatmap(object.list[[1]], pattern = "incoming", 
      signaling = all_pathways, title = names(object.list)[1], 
      width = 9, height = 11, color.heatmap = "GnBu")
ht6 = netAnalysis_signalingRole_heatmap(object.list[[2]], pattern ="incoming",
      signaling = all_pathways, title = names(object.list)[2], 
      width = 9, height = 11, color.heatmap = "GnBu")
incoming <- ComplexHeatmap::draw(ht5 + ht6, ht_gap = unit(0.5, "cm"))

```


# Figure 13A

```{r}
library(grid)
gb1 = grid.grabExpr(ComplexHeatmap::draw(ht3))
gb2 = grid.grabExpr(ComplexHeatmap::draw(ht4))
gb3 = grid.grabExpr(ComplexHeatmap::draw(ht5))
gb4 = grid.grabExpr(ComplexHeatmap::draw(ht6))

figure_13a <- cowplot::plot_grid(gb1, gb2, gb3, gb4, ncol = 2)
```


# Figure 13 B/C
```{r}

figure_13b <- netVisual_bubble(cellchat, sources.use =c(9, 2, 14, 7) , targets.use = c(6, 12, 16, 13),  
                 comparison = c(1, 2), angle.x = 45, signaling = "IL16") +theme(plot.margin = margin(l = 3, unit = "cm")) 

figure_13c <- netVisual_bubble(cellchat, sources.use =c(19, 20, 23) , targets.use = c(5, 6, 12, 15, 8, 22), comparison = c(1, 2), angle.x = 45, signaling = "CCL") +theme(plot.margin = margin(l = 3, unit = "cm"))
```

# Figure 13 All
```{r}
figure_13_all <- cowplot::plot_grid(figure_13a, figure_13b, figure_13c, ncol = 1, labels = c("A", "B", "C"), rel_heights = c(1,0.4, 0.4))
```

```{r}
ggsave(filename = paste0(fig_directory, "figure_13.pdf"), plot = figure_13_all, device = "pdf", height = 24, width = 20)

ggsave(filename = paste0(fig_directory, "figure_13.jpeg"), plot = figure_13_all, device = "jpeg", height = 24, width = 20)
```

# Overall Differential Interaction Strengths

Show differential number of interactions or interaction strength among 
different cell populations, red(increased signaling)/blue(decreased signaling)
```{r}
par(mfrow = c(1,2), xpd=TRUE)
netVisual_diffInteraction(cellchat, comparison = c(1, 2), measure = "count", 
                          weight.scale = T, arrow.size = 0.1)
netVisual_diffInteraction(cellchat, comparison = c(1, 2), measure = "weight", 
                          weight.scale = T, arrow.size = 0.1)
```

This is very complex, so if we want to merge our cells to get groups of cells
```
levels(object.list$Ca@meta[["CD_annot"]])

group.cellType <- c("CD4-lung", "CD4-mLns", "CD4-colon", rep("CD4_", 4), rep("DC", 4), rep("TC", 4))
group.cellType <- factor(group.cellType, levels = c("FIB", "DC", "TC"))
object.list <- lapply(object.list, function(x) {
                              mergeInteractions(x, group.cellType)})
cellchat <- mergeCellChat(object.list, add.names = names(object.list))

weight.max <- getMaxWeight(object.list, slot.name = c("idents", "net", "net"), 
                           attribute = c("idents", "count", "count.merged"))
```

```{r}
gg1 <- netVisual_heatmap(cellchat, comparison = c(1, 2), measure = "count", color.heatmap = c("blue", "red"))
gg2 <- netVisual_heatmap(cellchat, comparison = c(1, 2), measure = "weight", color.heatmap = c("blue", "red"))
gg1 + gg2
```

#  CCL5 Interaction Plot
```{r}
DefaultAssay(seurat_object) <- "RNA"
VlnPlot(seurat_object, group.by = "CD_annot", split.by = "Infection", features = c("Ccr4", "Ccl5"), stack = TRUE, flip = TRUE)
```

```{r}
par(mfrow = c(1,2), xpd=TRUE)
ht <- list()
for (i in 1:length(object.list)) {
  ht[[i]] <- netVisual_heatmap(object.list[[i]], signaling = c("CCL"), 
  title.name = paste("CCL", "signaling ",names(object.list)[i]),
  color.heatmap = "Reds")
}
ComplexHeatmap::draw(ht[[1]] + ht[[2]], ht_gap = unit(0.5, "cm"))
```

# IL16 Interaction Plot
```{r}
DefaultAssay(seurat_object) <- "RNA"
VlnPlot(seurat_object, group.by = "CD_annot", split.by = "Infection", features = c("Ccr4", "Ccl5"), stack = TRUE, flip = TRUE)
```

```{r}
par(mfrow = c(1,2), xpd=TRUE)
ht <- list()
for (i in 1:length(object.list)) {
  ht[[i]] <- netVisual_heatmap(object.list[[i]], signaling = c("IL16"), 
  title.name = paste("IL16", "signaling ",names(object.list)[i]),
  color.heatmap = "Reds")
}
ComplexHeatmap::draw(ht[[1]] + ht[[2]], ht_gap = unit(0.5, "cm"))
```

# IL2 Interaction Plot
```
#levels(cellchat.No@meta[["CD_annot"]])
# Using the KLRG1+ cells as the source for now
netVisual_bubble(object.list[[2]], sources.use =c(8) , targets.use = c(5,6,7,8,9,10,11,12,13,15, 19, 20),  
                 comparison = c(1, 2), angle.x = 45, signaling = "Il2 Ca") #, signaling = "IL2 Ca")

# Need to specify Il2 Ca, because it is not significant in No

netVisual_heatmap(object.list[[2]], signaling = c("IL2"))
```

# IFN-II Plot

```{r}
netVisual_bubble(object.list[[1]], sources.use =c(23) , targets.use = c(14, 16, 10, 6), angle.x = 45, signaling = "IFN-II")
```

```{r}
DefaultAssay(seurat_object) <- "RNA"
VlnPlot(seurat_object, group.by = "CD_annot", split.by = "Infection", features = c("Ifng", "Ifngr1", "Ifngr2"), stack = TRUE, flip = TRUE)
```

```{r}
netVisual_heatmap(object.list[[1]], signaling = c("IFN-II"))
```


